// @gsp-cmp/chgdr/esm2015/lib/chgdr-list-dialog-content/chgdr-list-dialog-content.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Injector, Input } from '@angular/core';
export class ChgdrListDialogContentComponent {
    /**
     * @param {?} injector
     */
    constructor(injector) {
        this.injector = injector;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
}
ChgdrListDialogContentComponent.decorators = [
    { type: Component, args: [{
                selector: 'lib-chgdr-list-dialog-content',
                template: "<chgdr-list [entityId]=\"entityId\" [dataId]=\"dataId\" style=\"padding: 0.75rem 0.875rem;height:100%\"></chgdr-list>",
                styles: [""]
            }] }
];
/** @nocollapse */
ChgdrListDialogContentComponent.ctorParameters = () => [
    { type: Injector }
];
ChgdrListDialogContentComponent.propDecorators = {
    entityId: [{ type: Input }],
    dataId: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    ChgdrListDialogContentComponent.prototype.entityId;
    /** @type {?} */
    ChgdrListDialogContentComponent.prototype.dataId;
    /**
     * @type {?}
     * @private
     */
    ChgdrListDialogContentComponent.prototype.injector;
}
