// @farris/devkit/esm5/lib/effector/repository_effector.js
import { Inject, Injectable, Injector } from '@angular/core';
import { BindingData } from '../binding-data/index';
import { EntityList } from '../entity/index';
import { NAMESPACE } from '../frame/tokens';
import { Repository } from '../repository/index';
/**
 * 实体副作用器
 * @description 将表达式计算结果赋值给实体属性
 */
var RepositoryEffector = /** @class */ (function () {
    function RepositoryEffector(injector, repository, namespace, bindingData) {
        this.injector = injector;
        this.repository = repository;
        this.namespace = namespace;
        this.bindingData = bindingData;
        this.ns = namespace;
    }
    RepositoryEffector.prototype.effect = function (path, value, options) {
        if (!options || !options.path) {
            throw new Error('repository effector 需要指定行信息。');
        }
        var propertyPath = options.path;
        var rowId = propertyPath[0] || this.bindingData.list.currentItem.primaryKeyValue;
        var entity = this.repository.entityCollection.getEntityById(rowId);
        if (rowId && !entity) {
            console.error("\u627E\u4E0D\u5230id\uFF1A" + rowId + "\u5BF9\u5E94\u7684\u5B9E\u4F53\uFF01");
            return;
        }
        // propertyPath like : [1, child1s, 1.1, child2s, 1.1.1,propName] or [1,udt,propName] or [1,prop]
        var propName = propertyPath.pop();
        var object = entity;
        for (var index = 1; index < propertyPath.length; index++) {
            var propertyName = propertyPath[index];
            if (object instanceof EntityList) {
                object = object.get(propertyName);
            }
            else {
                object = object[propertyName];
            }
        }
        if (object) {
            if (object[propName] !== value) {
                object[propName] = value;
            }
        }
        else {
            console.error("\u627E\u4E0D\u5230\u5B9E\u4F53\u5BF9\u5E94\u7684\u8DEF\u5F84\uFF1A" + propertyPath.push(propName));
        }
    };
    RepositoryEffector.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    RepositoryEffector.ctorParameters = function () { return [
        { type: Injector },
        { type: Repository },
        { type: undefined, decorators: [{ type: Inject, args: [NAMESPACE,] }] },
        { type: BindingData }
    ]; };
    return RepositoryEffector;
}());
export { RepositoryEffector };
