// @edp-bif/common-api/esm5/lib/content/impl/RtBillCreationProcessContent.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/content/impl/RtBillCreationProcessContent.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelJsonHelper } from '../../ref';
import { AbstractRtBillProcessContent } from './AbstractRtBillProcessContent';
import { RtBillCreationProcessEntity } from '../../entity/impl/RtBillCreationProcessEntity';
/**
 * 单据生单流内容类。
 */
var /**
 * 单据生单流内容类。
 */
RtBillCreationProcessContent = /** @class */ (function (_super) {
    tslib_1.__extends(RtBillCreationProcessContent, _super);
    function RtBillCreationProcessContent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    RtBillCreationProcessContent.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.formSettings = null;
        this.upBillTrackingSettings = null;
        this.downBillTrackingSettings = null;
    };
    /**
     * @return {?}
     */
    RtBillCreationProcessContent.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "formSettings", this.formSettings);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillTrackingSettings", this.upBillTrackingSettings);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillTrackingSettings", this.downBillTrackingSettings);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    RtBillCreationProcessContent.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.formSettings = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "formSettings");
        this.upBillTrackingSettings = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillTrackingSettings");
        this.downBillTrackingSettings = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillTrackingSettings");
    };
    /**
     * @protected
     * @return {?}
     */
    RtBillCreationProcessContent.prototype.getEntityClass = /**
     * @protected
     * @return {?}
     */
    function () {
        return RtBillCreationProcessEntity;
    };
    return RtBillCreationProcessContent;
}(AbstractRtBillProcessContent));
/**
 * 单据生单流内容类。
 */
export { RtBillCreationProcessContent };
if (false) {
    /** @type {?} */
    RtBillCreationProcessContent.prototype.formSettings;
    /** @type {?} */
    RtBillCreationProcessContent.prototype.upBillTrackingSettings;
    /** @type {?} */
    RtBillCreationProcessContent.prototype.downBillTrackingSettings;
}
