// @gsp-cmp/chgdr/esm2015/lib/model/operate-type.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class OperateType {
    /**
     * @private
     * @param {?} ordinal
     * @param {?} code
     * @param {?} name
     */
    constructor(ordinal, code, name) {
        this.ordinal = ordinal;
        this.code = code;
        this.name = name;
    }
    /**
     * @return {?}
     */
    getOrdinal() {
        return this.ordinal;
    }
    /**
     * @return {?}
     */
    getCode() {
        return this.code;
    }
    /**
     * @return {?}
     */
    getName() {
        return this.name;
    }
    /**
     * @return {?}
     */
    static values() {
        return [OperateType.ADD, OperateType.MODIFY, OperateType.DELETE];
    }
    /**
     * @return {?}
     */
    toJson() {
        return this.code;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    static parse(obj) {
        if (obj === undefined || obj === null) {
            return null;
        }
        if (obj instanceof OperateType) {
            return obj;
        }
        if (typeof obj === "string") {
            return OperateType.values().find((/**
             * @param {?} ot
             * @return {?}
             */
            ot => ot.code == obj));
        }
        if (typeof obj === "number") {
            return OperateType.values().find((/**
             * @param {?} ot
             * @return {?}
             */
            ot => ot.ordinal == obj));
        }
        if (!!obj.code) {
            return OperateType.values().find((/**
             * @param {?} ot
             * @return {?}
             */
            ot => ot.code == obj.code));
        }
        return null;
    }
}
/**
 * 新增
 */
OperateType.ADD = new OperateType(0, "ADD", "新增");
/**
 * 修改
 */
OperateType.MODIFY = new OperateType(1, "MODIFY", "修改");
/**
 * 删除
 */
OperateType.DELETE = new OperateType(2, "DELETE", "删除");
if (false) {
    /**
     * 新增
     * @type {?}
     */
    OperateType.ADD;
    /**
     * 修改
     * @type {?}
     */
    OperateType.MODIFY;
    /**
     * 删除
     * @type {?}
     */
    OperateType.DELETE;
    /** @type {?} */
    OperateType.prototype.ordinal;
    /** @type {?} */
    OperateType.prototype.code;
    /** @type {?} */
    OperateType.prototype.name;
}
