// @edp-bif/runtime-api/esm5/lib/entity/form-rest/FormBillCreationProcessExecutorResult.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/form-rest/FormBillCreationProcessExecutorResult.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { JsonConvertableImpl, BpmnModelJsonHelper } from "@edp-aif/common-api";
import { ConditionResult } from '../ConditionResult';
import { RpcCreationRuleExecutorResult } from '@edp-aif/runtime-api';
var FormBillCreationProcessExecutorResult = /** @class */ (function (_super) {
    tslib_1.__extends(FormBillCreationProcessExecutorResult, _super);
    function FormBillCreationProcessExecutorResult() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    FormBillCreationProcessExecutorResult.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.conditionResult = null;
        this.creationRuleExecutorResult = null;
    };
    /**
     * @return {?}
     */
    FormBillCreationProcessExecutorResult.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "conditionResult", this.conditionResult);
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleExecutorResult", this.creationRuleExecutorResult);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    FormBillCreationProcessExecutorResult.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.conditionResult = BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "conditionResult", ConditionResult);
        this.creationRuleExecutorResult = BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "creationRuleExecutorResult", RpcCreationRuleExecutorResult);
    };
    return FormBillCreationProcessExecutorResult;
}(JsonConvertableImpl));
export { FormBillCreationProcessExecutorResult };
if (false) {
    /** @type {?} */
    FormBillCreationProcessExecutorResult.prototype.conditionResult;
    /** @type {?} */
    FormBillCreationProcessExecutorResult.prototype.creationRuleExecutorResult;
}
