// @farris/devkit/esm5/lib/message/messagemanager.js
var MessageManager = /** @class */ (function () {
    // private parent:Window;
    // private children = new Map<string,Window>();
    function MessageManager() {
    }
    /**
     * 获取当前window上的MessageManager对象
     */
    MessageManager.getMessager = function () {
        if (!window['igix_messagemanager']) {
            window['igix_messagemanager'] = new MessageManager();
        }
        return window['igix_messagemanager'];
    };
    // addParentTarget(target: Window){
    //     this.parent = target;
    // }
    // addChildTarget(key: string,target: Window){
    //     this.children.set(key,target);
    // }
    MessageManager.prototype.listner = function (callbackfn) {
        window.addEventListener('message', function (e) {
            var data = e.data;
            data.sender = e.source;
            callbackfn(data);
        });
    };
    // listnerMessage(targetId: string,callbackfn: (e: any) => any){
    //     const win = this.children.get(targetId);
    //     if(!win){
    //         throw Error('This TargetId Can Not Find Window Object');
    //     }
    //     win.addEventListener('message',async function(e){
    //         await callbackfn(e);
    //     });
    // }
    MessageManager.prototype.send = function (msg) {
        // const win = this.children.get(targetId);
        var win;
        if (msg.target) {
            win = msg.target;
            delete msg.target;
        }
        else {
            win = window;
        }
        win.postMessage(msg, '*');
    };
    MessageManager.prototype.setAppContext = function (appContext) {
        this.curAppContext = appContext;
    };
    MessageManager.prototype.getAppContext = function () {
        return this.curAppContext;
    };
    return MessageManager;
}());
export { MessageManager };
