// @gsp-cmp/chgdr/esm5/lib/model/operate-type.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var OperateType = /** @class */ (function () {
    function OperateType(ordinal, code, name) {
        this.ordinal = ordinal;
        this.code = code;
        this.name = name;
    }
    /**
     * @return {?}
     */
    OperateType.prototype.getOrdinal = /**
     * @return {?}
     */
    function () {
        return this.ordinal;
    };
    /**
     * @return {?}
     */
    OperateType.prototype.getCode = /**
     * @return {?}
     */
    function () {
        return this.code;
    };
    /**
     * @return {?}
     */
    OperateType.prototype.getName = /**
     * @return {?}
     */
    function () {
        return this.name;
    };
    /**
     * @return {?}
     */
    OperateType.values = /**
     * @return {?}
     */
    function () {
        return [OperateType.ADD, OperateType.MODIFY, OperateType.DELETE];
    };
    /**
     * @return {?}
     */
    OperateType.prototype.toJson = /**
     * @return {?}
     */
    function () {
        return this.code;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    OperateType.parse = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        if (obj === undefined || obj === null) {
            return null;
        }
        if (obj instanceof OperateType) {
            return obj;
        }
        if (typeof obj === "string") {
            return OperateType.values().find((/**
             * @param {?} ot
             * @return {?}
             */
            function (ot) { return ot.code == obj; }));
        }
        if (typeof obj === "number") {
            return OperateType.values().find((/**
             * @param {?} ot
             * @return {?}
             */
            function (ot) { return ot.ordinal == obj; }));
        }
        if (!!obj.code) {
            return OperateType.values().find((/**
             * @param {?} ot
             * @return {?}
             */
            function (ot) { return ot.code == obj.code; }));
        }
        return null;
    };
    /**
     * 新增
     */
    OperateType.ADD = new OperateType(0, "ADD", "新增");
    /**
     * 修改
     */
    OperateType.MODIFY = new OperateType(1, "MODIFY", "修改");
    /**
     * 删除
     */
    OperateType.DELETE = new OperateType(2, "DELETE", "删除");
    return OperateType;
}());
export { OperateType };
if (false) {
    /**
     * 新增
     * @type {?}
     */
    OperateType.ADD;
    /**
     * 修改
     * @type {?}
     */
    OperateType.MODIFY;
    /**
     * 删除
     * @type {?}
     */
    OperateType.DELETE;
    /** @type {?} */
    OperateType.prototype.ordinal;
    /** @type {?} */
    OperateType.prototype.code;
    /** @type {?} */
    OperateType.prototype.name;
}
