// @farris/discussion-group/esm2015/lib/discussion-group.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule, registerLocaleData } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { SafeHtmlPipe } from './pipe/safe-html.pipe';
import { DiscussionListComponent } from './discussion-list/discussion-list.component';
import { DiscussionEditorComponent } from './discussion-editor/discussion-editor.component';
import { FarrisDialogModule } from '@farris/ui-dialog';
import { FarrisCommonModule } from '@farris/ui-common';
import { TreeModule } from '@farris/ui-tree';
import { ModalModule, BsModalService } from '@farris/ui-modal';
import { LoadingModule } from '@farris/ui-loading';
import { LocaleModule } from '@farris/ui-locale';
import { ComponentLoaderFactory } from '@farris/ui-modal/component-loader';
import { PositioningService } from '@farris/ui-modal/positioning';
import { PerfectScrollbarModule } from '@farris/ui-perfect-scrollbar';
import { DiscussionPersonnelComponent } from './discussion-personnel/discussion-personnel.component';
import { PaginationModule } from '@farris/ui-pagination';
import { DiscussionListService } from './discussion-list/discussion-list.service';
import { DiscussionListDirective } from './discussion-list/discussionList.directive';
import { FFileUploadModule } from '@farris/extend-file-upload';
import { NotifyModule } from '@farris/ui-notify';
import { InputGroupModule } from '@farris/ui-input-group';
import zh from '@angular/common/locales/zh';
import { HttpClientModule } from '@angular/common/http';
import { AvatarHtmlPipe, GetOrgTextPipe } from './pipe/controlType.pipe';
registerLocaleData(zh);
export class DiscussionGroupModule {
}
DiscussionGroupModule.decorators = [
    { type: NgModule, args: [{
                declarations: [DiscussionEditorComponent, DiscussionListComponent, DiscussionPersonnelComponent, SafeHtmlPipe, AvatarHtmlPipe, GetOrgTextPipe, DiscussionListDirective],
                imports: [
                    CommonModule,
                    FormsModule,
                    ModalModule,
                    FarrisDialogModule,
                    FarrisCommonModule,
                    LoadingModule,
                    PerfectScrollbarModule,
                    TreeModule,
                    PaginationModule,
                    FFileUploadModule,
                    NotifyModule,
                    HttpClientModule,
                    LocaleModule.forRoot(),
                    InputGroupModule
                ],
                providers: [
                    BsModalService,
                    ComponentLoaderFactory,
                    PositioningService,
                    DiscussionListService
                ],
                exports: [DiscussionEditorComponent, DiscussionListComponent, DiscussionPersonnelComponent, SafeHtmlPipe, AvatarHtmlPipe, GetOrgTextPipe, DiscussionListDirective],
                entryComponents: [DiscussionPersonnelComponent]
            },] }
];
