// @edp-aif/common-api/esm5/lib/billkit/BillSchemaFieldBase.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelJsonHelper } from '../BpmnModelJsonHelper';
import { KitField } from './KitField';
import { DataType } from './DataType';
/**
 * @abstract
 */
var /**
 * @abstract
 */
BillSchemaFieldBase = /** @class */ (function (_super) {
    tslib_1.__extends(BillSchemaFieldBase, _super);
    function BillSchemaFieldBase() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.dataType = DataType.Unknow;
        _this.extendProperties = {};
        _this.length = 0;
        _this.precision = 0;
        _this.scale = 0;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    BillSchemaFieldBase.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.fieldType = null;
        this.dataType = DataType.Unknow;
        this.caption = null;
        this.extendProperties = {};
        this.length = 0;
        this.precision = 0;
        this.scale = 0;
    };
    /**
     * @return {?}
     */
    BillSchemaFieldBase.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "type", this.fieldType);
        if (this.dataType != DataType.String) {
            BpmnModelJsonHelper.SetAttributeValue(obj, "data", this.dataType);
        }
        BpmnModelJsonHelper.SetAttributeValue(obj, "cap", this.caption);
        BpmnModelJsonHelper.SetAttributeValue(obj, "props", Object.assign({}, this.extendProperties));
        BpmnModelJsonHelper.SetAttributeValue(obj, "length", this.length);
        BpmnModelJsonHelper.SetAttributeValue(obj, "precision", this.precision);
        BpmnModelJsonHelper.SetAttributeValue(obj, "scale", this.scale);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    BillSchemaFieldBase.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.fieldType = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "type");
        /** @type {?} */
        var data = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "data");
        if (data == null) {
            this.dataType = DataType.String;
        }
        else {
            this.dataType = DataType[data];
        }
        this.dataType = DataType[BpmnModelJsonHelper.GetAttributeValueAsString(obj, "data")];
        this.caption = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "cap");
        this.extendProperties = BpmnModelJsonHelper.GetAttributeValue(obj, "props");
        this.length = BpmnModelJsonHelper.GetAttributeValueAsNumber(obj, "length");
        this.precision = BpmnModelJsonHelper.GetAttributeValueAsNumber(obj, "precision");
        this.scale = BpmnModelJsonHelper.GetAttributeValueAsNumber(obj, "scale");
    };
    return BillSchemaFieldBase;
}(KitField));
/**
 * @abstract
 */
export { BillSchemaFieldBase };
if (false) {
    /** @type {?} */
    BillSchemaFieldBase.prototype.fieldType;
    /** @type {?} */
    BillSchemaFieldBase.prototype.dataType;
    /** @type {?} */
    BillSchemaFieldBase.prototype.caption;
    /** @type {?} */
    BillSchemaFieldBase.prototype.extendProperties;
    /** @type {?} */
    BillSchemaFieldBase.prototype.attachedField;
    /** @type {?} */
    BillSchemaFieldBase.prototype.length;
    /** @type {?} */
    BillSchemaFieldBase.prototype.precision;
    /** @type {?} */
    BillSchemaFieldBase.prototype.scale;
}
