// @gsp-cmp/chgdr/esm2015/lib/chgdr-list/chgdr-list.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, HostBinding, Injector, Input } from '@angular/core';
import { MessagerService } from '@farris/ui-messager';
import { NotifyService } from '@farris/ui-notify';
import { ChgdrService } from '../service/chgdr.service';
import { ChangeDataQueryParam } from '../model/change-data-query-param';
import { ChangeDataQueryResult } from '../model/change-data-query-result';
import { ErrorUtil } from '../util/error.util';
import { TimeUtil } from '../util/time.util';
export class ChgdrListComponent {
    /**
     * @param {?} chgdrService
     * @param {?} messageService
     * @param {?} notifyService
     * @param {?} injector
     */
    constructor(chgdrService, messageService, notifyService, injector) {
        this.chgdrService = chgdrService;
        this.messageService = messageService;
        this.notifyService = notifyService;
        this.injector = injector;
        this.baseCls = true;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        /** @type {?} */
        let chgdrs = new ChangeDataQueryResult();
        chgdrs.total = 0;
        chgdrs.pageSize = 20;
        chgdrs.pageIndex = 1;
        chgdrs.headers = [];
        this.chgdrs = chgdrs;
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (!this.changeTimeStart && !this.changeTimeEnd) {
            //未录入变更时间范围，则查询最近一周的
            this.changeTimeEnd = TimeUtil.getEndTimeInDay(new Date());
            this.changeTimeStart = TimeUtil.getStartTimeInDay(new Date(new Date().getTime() - 7 * 24 * 60 * 60 * 1000));
        }
        else if (!this.changeTimeEnd) {
            //如果未设置截至时间，则与开始时间相同
            this.changeTimeEnd = TimeUtil.getEndTimeInDay(this.changeTimeStart);
        }
        else if (!this.changeTimeStart) {
            //如果未设置开始时间，则与截至时间相同
            this.changeTimeStart = TimeUtil.getEndTimeInDay(this.changeTimeEnd);
        }
        if (!!this.entityId) {
            /** @type {?} */
            let param = this.buildQueryParam();
            this.doQuery(param);
        }
    }
    /**
     * @return {?}
     */
    query() {
        /** @type {?} */
        let param = this.buildQueryParam();
        this.doQuery(param);
    }
    /**
     * @private
     * @return {?}
     */
    buildQueryParam() {
        /** @type {?} */
        let param = new ChangeDataQueryParam();
        param.pageIndex = 1;
        param.pageSize = 20;
        param.entityId = this.entityId;
        param.dataId = this.dataId;
        param.changeTimeStart = this.changeTimeStart;
        param.changeTimeEnd = this.changeTimeEnd;
        return param;
    }
    /**
     * @private
     * @param {?} param
     * @return {?}
     */
    doQuery(param) {
        this.chgdrService.queryChangeDataHeader(param).subscribe((/**
         * @param {?} data
         * @return {?}
         */
        data => {
            this.currentQueryParam = param;
            this.chgdrs = data;
        }), (/**
         * @param {?} error
         * @return {?}
         */
        error => {
            console.error("查询业务变更日志出错", error);
            this.messageService.error("查询业务变更日志出错：" + ErrorUtil.getErrorMessage(error));
        }));
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onPageChanged($event) {
        if (!$event) {
            return;
        }
        if ($event.pageIndex) {
            this.currentQueryParam.pageIndex = $event.pageIndex;
        }
        if ($event.pageSize) {
            this.currentQueryParam.pageSize = $event.pageSize;
        }
        this.doQuery(this.currentQueryParam);
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onPageSizeChanged($event) {
        this.currentQueryParam.pageSize = $event;
        this.doQuery(this.currentQueryParam);
    }
}
ChgdrListComponent.decorators = [
    { type: Component, args: [{
                selector: 'chgdr-list',
                template: "<div class=\"farris-form farris-form-controls-inline\" style=\"background: #fff;\">\n  <form class=\"f-section-formgroup\" style=\"padding-top: 10px;\">\n    <div class=\"f-section-formgroup-inputs\">\n      <div class=\"col-12 row\">\n        <div class=\"farris-group-wrap col-5\">\n          <div class=\"form-group farris-form-group\">\n            <label class=\"col-form-label\">\n              <span class=\"farris-label-info text-danger\">*</span><span class=\"farris-label-text\">\u5F00\u59CB\u65F6\u95F4</span>\n            </label>\n            <div class=\"farris-input-wrap\">\n              <farris-datepicker [(ngModel)]=\"changeTimeStart\" name=\"changeTimeStart\"></farris-datepicker>\n            </div>\n          </div>\n        </div>\n        <div class=\"farris-group-wrap col-5\">\n          <div class=\"form-group farris-form-group\">\n            <label class=\"col-form-label\">\n              <span class=\"farris-label-info text-danger\">*</span><span class=\"farris-label-text\">\u7ED3\u675F\u65F6\u95F4</span>\n            </label>\n            <div class=\"farris-input-wrap\">\n              <farris-datepicker [(ngModel)]=\"changeTimeEnd\" name=\"changeTimeEnd\"></farris-datepicker>\n            </div>\n          </div>\n        </div>\n        <div class=\"farris-group-wrap col-2\">\n          <div class=\"form-group farris-form-group\">\n            <label class=\"col-form-label\">\n              <button class=\"btn btn-primary mr-2\" (click)=\"query()\">\u67E5\u8BE2</button>\n            </label>\n            <div class=\"farris-input-wrap\">\n            </div>\n          </div>\n        </div>\n      </div>\n    </div>\n  </form>\n</div>\n\n<lib-chgdr-header-datagrid\n  [chgdrs]=\"chgdrs\" \n  (pageChanged)=\"onPageChanged($event)\" \n  (pageSizeChanged)=\"onPageSizeChanged($event)\"\n  [showDataId]=\"false\"\n></lib-chgdr-header-datagrid>\n\n<!-- <div class=\"f-utils-fill\">\n  <farris-datagrid #chgdrGrid idField=\"id\" [columns]=\"columns\" [data]=\"chgdrs.headers\"\n  showLineNumber=\"true\" [fit]=\"true\" [fitColumns]=\"true\" \n  [pagination]=\"true\" [pageSize]=\"20\" [showPageList]=\"false\" [pageIndex]=\"chgdrs.pageIndex\" [total]=\"chgdrs.total\"\n  (pageChanged)=\"onPageChanged($event)\" (pageSizeChanged)=\"onPageSizeChanged($event)\"\n  >\n  </farris-datagrid>\n  <ng-template #opCell let-ctx>\n    <a href=\"javascript: void(0);\" (click)=\"showChangeDetail(ctx)\">\u53D8\u66F4\u8BE6\u60C5</a>\n  </ng-template>\n</div> -->\n\n<!-- \u67E5\u770B\u53D8\u66F4\u65E5\u5FD7\u7684\u5F39\u7A97 -->\n<!-- <farris-dialog #chgdrInfoDialog\n  title=\"\u53D8\u66F4\u65E5\u5FD7\u8BE6\u60C5\"\n  [width]=\"1100\"\n  [height]=\"600\"\n  [showMaxButton]=\"true\"\n  [resizable]=\"true\"\n  [showButtons]=\"false\"\n>\n  <chgdr-data-viewer [data]=\"currentChgdrData\"></chgdr-data-viewer>\n</farris-dialog> -->",
                styles: [""]
            }] }
];
/** @nocollapse */
ChgdrListComponent.ctorParameters = () => [
    { type: ChgdrService },
    { type: MessagerService },
    { type: NotifyService },
    { type: Injector }
];
ChgdrListComponent.propDecorators = {
    baseCls: [{ type: HostBinding, args: ['class.f-utils-fill-flex-column',] }],
    entityId: [{ type: Input }],
    dataId: [{ type: Input }],
    changeTimeStart: [{ type: Input }],
    changeTimeEnd: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    ChgdrListComponent.prototype.baseCls;
    /** @type {?} */
    ChgdrListComponent.prototype.entityId;
    /** @type {?} */
    ChgdrListComponent.prototype.dataId;
    /** @type {?} */
    ChgdrListComponent.prototype.changeTimeStart;
    /** @type {?} */
    ChgdrListComponent.prototype.changeTimeEnd;
    /**
     * @type {?}
     * @private
     */
    ChgdrListComponent.prototype.currentQueryParam;
    /** @type {?} */
    ChgdrListComponent.prototype.chgdrs;
    /**
     * @type {?}
     * @private
     */
    ChgdrListComponent.prototype.chgdrService;
    /**
     * @type {?}
     * @private
     */
    ChgdrListComponent.prototype.messageService;
    /**
     * @type {?}
     * @private
     */
    ChgdrListComponent.prototype.notifyService;
    /**
     * @type {?}
     * @private
     */
    ChgdrListComponent.prototype.injector;
}
