// @farris/ui-field-group/esm2015/lib/form/form.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ContentChildren, QueryList } from '@angular/core';
import { RowComponent } from '../row/row.component';
export class FormComponent {
    constructor() {
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        this.arr.forEach((/**
         * @param {?} item
         * @param {?} index
         * @param {?} array
         * @return {?}
         */
        (item, index, array) => {
            console.log(item);
        }));
    }
}
FormComponent.decorators = [
    { type: Component, args: [{
                selector: 'farris-form',
                template: "<ng-content></ng-content>",
                styles: [""]
            }] }
];
/** @nocollapse */
FormComponent.ctorParameters = () => [];
FormComponent.propDecorators = {
    arr: [{ type: ContentChildren, args: [RowComponent,] }]
};
if (false) {
    /** @type {?} */
    FormComponent.prototype.arr;
}
