// @farris/ui-common/esm5/lib/directives/resize-observer.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Directive, Output, ElementRef, EventEmitter, Input } from '@angular/core';
import ResizeObserver from 'resize-observer-polyfill';
/** @type {?} */
var entriesMap = new WeakMap();
var ɵ0 = /**
 * @param {?} entries
 * @return {?}
 */
function (entries) {
    var e_1, _a;
    try {
        for (var entries_1 = tslib_1.__values(entries), entries_1_1 = entries_1.next(); !entries_1_1.done; entries_1_1 = entries_1.next()) {
            var entry = entries_1_1.value;
            if (entriesMap.has(entry.target)) {
                /** @type {?} */
                var comp = entriesMap.get(entry.target);
                comp._resizeCallback(entry);
            }
        }
    }
    catch (e_1_1) { e_1 = { error: e_1_1 }; }
    finally {
        try {
            if (entries_1_1 && !entries_1_1.done && (_a = entries_1.return)) _a.call(entries_1);
        }
        finally { if (e_1) throw e_1.error; }
    }
};
/** @type {?} */
var ro = new ResizeObserver((ɵ0));
var ResizeObserverDirective = /** @class */ (function () {
    function ResizeObserverDirective(el) {
        this.el = el;
        this.resize = new EventEmitter();
    }
    /**
     * @param {?} entry
     * @return {?}
     */
    ResizeObserverDirective.prototype._resizeCallback = /**
     * @param {?} entry
     * @return {?}
     */
    function (entry) {
        entry['bindintClientRect'] = entry.target.getBoundingClientRect();
        // console.log(entry.contentRect, entry.bindintClientRect);
        this.resize.emit(entry);
    };
    /**
     * @return {?}
     */
    ResizeObserverDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var target = this.target ? this.target.nativeElement : this.el.nativeElement;
        entriesMap.set(target, this);
        ro.observe(target);
    };
    /**
     * @return {?}
     */
    ResizeObserverDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var target = this.el.nativeElement;
        ro.unobserve(target);
        entriesMap.delete(target);
    };
    ResizeObserverDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[resizeObserver]'
                },] }
    ];
    /** @nocollapse */
    ResizeObserverDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    ResizeObserverDirective.propDecorators = {
        target: [{ type: Input, args: ['resizeObserver',] }],
        resize: [{ type: Output }]
    };
    return ResizeObserverDirective;
}());
export { ResizeObserverDirective };
if (false) {
    /** @type {?} */
    ResizeObserverDirective.prototype.target;
    /** @type {?} */
    ResizeObserverDirective.prototype.resize;
    /** @type {?} */
    ResizeObserverDirective.prototype.el;
}
export { ɵ0 };
