// @gsp-cmp/chgdr/esm5/lib/service/chgdr-list-ui.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, ComponentFactoryResolver, Injector } from '@angular/core';
import { MessagerService } from '@farris/ui-messager';
import { BsModalService } from '@farris/ui-modal';
import { ChgdrListDialogContentComponent } from '../chgdr-list-dialog-content/chgdr-list-dialog-content.component';
import * as i0 from "@angular/core";
import * as i1 from "@farris/ui-modal";
import * as i2 from "@farris/ui-messager";
var ChgdrListUiService = /** @class */ (function () {
    function ChgdrListUiService(resolver, injector, modalService, messagerService) {
        this.resolver = resolver;
        this.injector = injector;
        this.modalService = modalService;
        this.messagerService = messagerService;
    }
    /**
     * @param {?} entityId
     * @param {?} dataId
     * @return {?}
     */
    ChgdrListUiService.prototype.showDialog = /**
     * @param {?} entityId
     * @param {?} dataId
     * @return {?}
     */
    function (entityId, dataId) {
        return this.create(entityId, dataId);
    };
    /**
     * @private
     * @param {?} entityId
     * @param {?} dataId
     * @return {?}
     */
    ChgdrListUiService.prototype.create = /**
     * @private
     * @param {?} entityId
     * @param {?} dataId
     * @return {?}
     */
    function (entityId, dataId) {
        if (!entityId) {
            console.error("创建业务变更日志列表组件失败，业务实体不能为空，请检查", entityId, dataId);
            return;
        }
        /** @type {?} */
        var fac = this.resolver.resolveComponentFactory(ChgdrListDialogContentComponent);
        /** @type {?} */
        var cmpRef = fac.create(this.injector);
        /** @type {?} */
        var modalConfig = {
            title: '变更日志列表',
            width: 1000,
            height: 550,
            resizable: true,
            showButtons: false,
            showMaxButton: true
        };
        cmpRef.instance.entityId = entityId;
        cmpRef.instance.dataId = dataId;
        /** @type {?} */
        var dialog = this.modalService.show(cmpRef, modalConfig);
    };
    ChgdrListUiService.decorators = [
        { type: Injectable, args: [{ providedIn: 'root' },] }
    ];
    /** @nocollapse */
    ChgdrListUiService.ctorParameters = function () { return [
        { type: ComponentFactoryResolver },
        { type: Injector },
        { type: BsModalService },
        { type: MessagerService }
    ]; };
    /** @nocollapse */ ChgdrListUiService.ngInjectableDef = i0.defineInjectable({ factory: function ChgdrListUiService_Factory() { return new ChgdrListUiService(i0.inject(i0.ComponentFactoryResolver), i0.inject(i0.INJECTOR), i0.inject(i1.BsModalService), i0.inject(i2.MessagerService)); }, token: ChgdrListUiService, providedIn: "root" });
    return ChgdrListUiService;
}());
export { ChgdrListUiService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ChgdrListUiService.prototype.resolver;
    /**
     * @type {?}
     * @private
     */
    ChgdrListUiService.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    ChgdrListUiService.prototype.modalService;
    /**
     * @type {?}
     * @private
     */
    ChgdrListUiService.prototype.messagerService;
}
