// @edp-bif/common-api/esm2015/lib/slim-entity/impl/RtSlimBillCreationProcessEntity.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/slim-entity/impl/RtSlimBillCreationProcessEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModelJsonHelper } from "../../ref";
import { AbstractRtSlimBillProcessEntity } from "./AbstractRtSlimBillProcessEntity";
/**
 * 运行时简化生单流实体类。
 */
export class RtSlimBillCreationProcessEntity extends AbstractRtSlimBillProcessEntity {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.formSettingsId = null;
        this.creationRuleId = null;
        this.downBillVoProxyMetadataId = null;
        this.downBillVoProxyId = null;
        this.upBillTrackingSettingsId = null;
        this.downBillTrackingSettingsId = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "formSettingsId", this.formSettingsId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleId", this.creationRuleId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoProxyMetadataId", this.downBillVoProxyMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoProxyId", this.downBillVoProxyId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillTrackingSettingsId", this.upBillTrackingSettingsId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillTrackingSettingsId", this.downBillTrackingSettingsId);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.formSettingsId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "formSettingsId");
        this.creationRuleId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleId");
        this.downBillVoProxyMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoProxyMetadataId");
        this.downBillVoProxyId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoProxyId");
        this.upBillTrackingSettingsId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillTrackingSettingsId");
        this.downBillTrackingSettingsId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillTrackingSettingsId");
    }
}
if (false) {
    /** @type {?} */
    RtSlimBillCreationProcessEntity.prototype.formSettingsId;
    /** @type {?} */
    RtSlimBillCreationProcessEntity.prototype.creationRuleId;
    /** @type {?} */
    RtSlimBillCreationProcessEntity.prototype.downBillVoProxyMetadataId;
    /** @type {?} */
    RtSlimBillCreationProcessEntity.prototype.downBillVoProxyId;
    /** @type {?} */
    RtSlimBillCreationProcessEntity.prototype.upBillTrackingSettingsId;
    /** @type {?} */
    RtSlimBillCreationProcessEntity.prototype.downBillTrackingSettingsId;
}
