// @farris/command-services/esm5/lib/form-notify.service.js
import { Injectable, Optional } from '@angular/core';
import { NotifyService } from '@farris/ui-notify';
import { LanguageService } from './languag.service';
/**
 * 通知服务
 * @scope FormModule
 * @description
 * 消息提示使用策略：成功使用success，失败使用error，警告使用warning，其他场景使用info
 */
var FormNotifyService = /** @class */ (function () {
    /**
     * 构造函数
     * 注入@farris/ui的NotifyService
     */
    function FormNotifyService(notifyService, languageService) {
        this.notifyService = notifyService;
        this.languageService = languageService;
        if (!languageService) {
            this.languageService = LanguageService.getInstance();
        }
    }
    /**
     * 信息提示
     * @param content 内容
     */
    FormNotifyService.prototype.default = function (content) {
        console.log("DEFAULT: " + content);
        return this.notifyService.default({
            title: this.languageService.notifyTitle,
            msg: content,
            timeout: 3000
        });
    };
    /**
     * 信息提示
     * @param content 内容
     */
    FormNotifyService.prototype.info = function (content, options) {
        console.log("INFO: " + content);
        var notifyOptions = {
            title: this.languageService.notifyTitle,
            msg: content,
            timeout: 3000
        };
        if (options && options.hideTitle) {
            delete notifyOptions.title;
        }
        return this.notifyService.info(notifyOptions);
    };
    /**
     * 成功提示
     * @param content 内容
     * @param options? 选项 {shorTitle: false}
     */
    FormNotifyService.prototype.success = function (content, options) {
        console.log("SUCCESS: " + content);
        var notifyOptions = {
            title: this.languageService.notifyTitle,
            msg: content,
            timeout: 3000
        };
        if (options && options.hideTitle) {
            delete notifyOptions.title;
        }
        this.notifyService.success(notifyOptions);
    };
    /**
     * 警告提示
     * @param content 内容
     * @param options 选项
     */
    FormNotifyService.prototype.warning = function (content, options) {
        var notifyOptions = {
            title: this.languageService.notifyTitle,
            msg: content,
            timeout: 3000
        };
        if (options && options.hideTitle) {
            delete notifyOptions.title;
        }
        this.notifyService.warning(notifyOptions);
    };
    /**
     * 错误提示
     * @param content 内容
     */
    FormNotifyService.prototype.error = function (content, options) {
        console.log("ERROR: " + content);
        var notifyOptions = {
            title: this.languageService.notifyTitle,
            msg: content,
            timeout: 3000
        };
        if (options && options.hideTitle) {
            delete notifyOptions.title;
        }
        this.notifyService.error(notifyOptions);
    };
    FormNotifyService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    FormNotifyService.ctorParameters = function () { return [
        { type: NotifyService },
        { type: LanguageService, decorators: [{ type: Optional }] }
    ]; };
    return FormNotifyService;
}());
export { FormNotifyService };
