// @farris/ui-context-menu/esm2015/lib/context-menu.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FarrisContextMenuComponent } from './context-menu.component';
import { FarrisContextMenuDirective } from './context-menu.directive';
import { FarrisContextMenuService } from './context-menu.service';
export class FarrisContextMenuModule {
}
FarrisContextMenuModule.decorators = [
    { type: NgModule, args: [{
                declarations: [
                    FarrisContextMenuComponent,
                    FarrisContextMenuDirective
                ],
                imports: [
                    CommonModule
                ],
                exports: [
                    FarrisContextMenuComponent,
                    FarrisContextMenuDirective
                ],
                entryComponents: [
                    FarrisContextMenuComponent
                ],
                providers: [
                    FarrisContextMenuService
                ],
            },] }
];
