// @farris/ui-datagrid-settings/esm5/lib/simple-mode/simple-columns.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectorRef, Component, EventEmitter, HostBinding, Injector, Input, Output, ViewEncapsulation, HostListener, ViewChild, ElementRef } from '@angular/core';
import { LocaleService } from '@farris/ui-locale';
import { MessagerService } from '@farris/ui-messager';
var SimpleColumnsComponent = /** @class */ (function () {
    function SimpleColumnsComponent(injector, cd) {
        this.injector = injector;
        this.cd = cd;
        this.seartTypes = [];
        /**
         * 所有列
         */
        this.columns = [];
        /**
         * 显示的列
         */
        this.fields = [];
        /**
         * 点击侧边样式以外的区域隐藏面板。默认为 true
         */
        this.clickToHidden = true;
        this.top = 0;
        this.cls = 'f-datagrid-settings-simple-host';
        this.formatChange = new EventEmitter();
        this.fieldsChange = new EventEmitter();
        this.submit = new EventEmitter();
        this.closed = new EventEmitter();
        this.advanced = new EventEmitter();
        this.allCols = [];
        this.columnType = 'all';
        this.isEn = false;
        this.settings = {};
        this.localeSer = this.injector.get(LocaleService, null);
        this.messager = this.injector.get(MessagerService, null);
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    SimpleColumnsComponent.prototype.onClick = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        // console.log($event);
        $event.stopPropagation();
        /** @type {?} */
        var clsname = $event.target['className'] || '';
        if (clsname.indexOf('f-datagrid-settings-simple-host') > -1) {
            // 隐藏侧边栏；
            if (this.clickToHidden) {
                this.close();
            }
        }
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    SimpleColumnsComponent.prototype.onEscKeydown = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        $event.stopPropagation();
        if (!document.querySelector('modal-container')) {
            this.close();
        }
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    SimpleColumnsComponent.prototype.onEnterKeydown = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        $event.stopPropagation();
        if (!document.querySelector('modal-container')) {
            this.onSubmitClick();
        }
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    SimpleColumnsComponent.prototype.onControlGKeydown = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        if ($event.ctrlKey && !document.querySelector('modal-container')) {
            $event.preventDefault();
            $event.stopPropagation();
            if ($event.key.toLowerCase() === 'g') {
                this.showAdvanced($event);
            }
            else if ($event.key.toLowerCase() === 'r') {
                this.onSettingReset();
            }
            else if ($event.key.toLowerCase() === 's') {
                this.searchInput.nativeElement.focus();
            }
        }
    };
    /**
     * @return {?}
     */
    SimpleColumnsComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.allCols = this.columns[0];
        this.fields = this.getViewColumns();
        this.container.nativeElement.focus();
    };
    /**
     * @private
     * @return {?}
     */
    SimpleColumnsComponent.prototype.getViewColumns = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        // 多表头
        /** @type {?} */
        var visibleCls = [];
        if (this.columns.length > 1) {
            this.columns.forEach((/**
             * @param {?} cols
             * @return {?}
             */
            function (cols) {
                visibleCls = visibleCls.concat(_this._getViewColumns(cols));
            }));
        }
        else {
            visibleCls = this.columns[0].filter((/**
             * @param {?} n
             * @return {?}
             */
            function (n) { return (n.visible || n.visible === undefined); }));
        }
        return visibleCls.map((/**
         * @param {?} n
         * @return {?}
         */
        function (n) { return n.field; }));
    };
    /**
     * @private
     * @param {?} cols
     * @return {?}
     */
    SimpleColumnsComponent.prototype._getViewColumns = /**
     * @private
     * @param {?} cols
     * @return {?}
     */
    function (cols) {
        return cols.filter((/**
         * @param {?} c
         * @return {?}
         */
        function (c) { return !c.colspan && (c.visible || c.visible === undefined); }));
    };
    // 根据列的类型进行筛选过滤，
    // all 所有列
    // visible 可见列
    // hidden 隐藏列
    // group 分组列
    // 根据列的类型进行筛选过滤，
    // all 所有列
    // visible 可见列
    // hidden 隐藏列
    // group 分组列
    /**
     * @param {?} ct
     * @param {?=} searchWords
     * @return {?}
     */
    SimpleColumnsComponent.prototype.filterColumns = 
    // 根据列的类型进行筛选过滤，
    // all 所有列
    // visible 可见列
    // hidden 隐藏列
    // group 分组列
    /**
     * @param {?} ct
     * @param {?=} searchWords
     * @return {?}
     */
    function (ct, searchWords) {
        var _this = this;
        if (searchWords === void 0) { searchWords = ''; }
        /** @type {?} */
        var cols = [];
        switch (ct) {
            case 'visible':
                cols = this.columns[0].filter((/**
                 * @param {?} n
                 * @return {?}
                 */
                function (n) { return _this.fields.indexOf(n.field) > -1; }));
                break;
            case 'hidden':
                cols = this.columns[0].filter((/**
                 * @param {?} n
                 * @return {?}
                 */
                function (n) { return _this.fields.indexOf(n.field) === -1; }));
                break;
            case 'all':
            default:
                cols = this.columns[0];
                break;
        }
        if (searchWords) {
            cols = cols.filter((/**
             * @param {?} c
             * @return {?}
             */
            function (c) { return c.title.toLowerCase().indexOf(searchWords) > -1; }));
        }
        this.allCols = cols;
    };
    /**
     * @private
     * @return {?}
     */
    SimpleColumnsComponent.prototype.onFieldsChanged = /**
     * @private
     * @return {?}
     */
    function () {
        this.fieldsChange.emit(this.fields);
    };
    /**
     * @param {?} $event
     * @param {?} col
     * @return {?}
     */
    SimpleColumnsComponent.prototype.onDblClick = /**
     * @param {?} $event
     * @param {?} col
     * @return {?}
     */
    function ($event, col) {
        // console.log('显示与隐藏切换事件', col);
        this.onCheckboxClick($event, col);
    };
    /**
     * @param {?} $event
     * @param {?} col
     * @return {?}
     */
    SimpleColumnsComponent.prototype.onCheckboxClick = /**
     * @param {?} $event
     * @param {?} col
     * @return {?}
     */
    function ($event, col) {
        var _this = this;
        // console.log('显示与隐藏切换事件', col);
        $event.stopPropagation();
        /** @type {?} */
        var index = this.fields.indexOf(col.field);
        if (index === -1) {
            this.fields.push(col.field);
        }
        else {
            this.fields.splice(index, 1);
        }
        this.columns[0].forEach((/**
         * @param {?} f
         * @return {?}
         */
        function (f) {
            f.visible = _this.fields.indexOf(f.field) > -1;
        }));
        this.fields = this.columns[0].filter((/**
         * @param {?} n
         * @return {?}
         */
        function (n) { return n.visible; })).map((/**
         * @param {?} n
         * @return {?}
         */
        function (n) { return n.field; }));
        this.onFieldsChanged();
    };
    /**
     * @param {?} $event
     * @param {?} ct
     * @return {?}
     */
    SimpleColumnsComponent.prototype.searchColumn = /**
     * @param {?} $event
     * @param {?} ct
     * @return {?}
     */
    function ($event, ct) {
        /** @type {?} */
        var txt = $event.target.value;
        this.filterColumns(ct, txt);
    };
    /**
     * @param {?} fieldInfo
     * @param {?} ct
     * @param {?} txt
     * @param {?} ulElRef
     * @return {?}
     */
    SimpleColumnsComponent.prototype.onFieldDrop = /**
     * @param {?} fieldInfo
     * @param {?} ct
     * @param {?} txt
     * @param {?} ulElRef
     * @return {?}
     */
    function (fieldInfo, ct, txt, ulElRef) {
        var _a;
        // console.log('drop', fieldInfo);
        var removedIndex = fieldInfo.removedIndex, addedIndex = fieldInfo.addedIndex;
        /** @type {?} */
        var moveItem = this.allCols[removedIndex];
        /** @type {?} */
        var targetItem = this.allCols[addedIndex];
        /** @type {?} */
        var moveItemIndex = this.columns[0].findIndex((/**
         * @param {?} n
         * @return {?}
         */
        function (n) { return n.field === moveItem.field; }));
        /** @type {?} */
        var targetItemIndex = this.columns[0].findIndex((/**
         * @param {?} n
         * @return {?}
         */
        function (n) { return n.field === targetItem.field; }));
        /** @type {?} */
        var removedItem = this.columns[0].splice(moveItemIndex, 1);
        (_a = this.columns[0]).splice.apply(_a, tslib_1.__spread([targetItemIndex, 0], removedItem));
        this.fields = this.columns[0].filter((/**
         * @param {?} n
         * @return {?}
         */
        function (n) { return n.visible; })).map((/**
         * @param {?} n
         * @return {?}
         */
        function (n) { return n.field; }));
        this.onFieldsChanged();
        this.filterColumns(ct, txt);
    };
    /**
     * 0: 1个也没有选中； 1: 全部选中； 2: 有选中的也有没选中的
     */
    /**
     * 0: 1个也没有选中； 1: 全部选中； 2: 有选中的也有没选中的
     * @private
     * @return {?}
     */
    SimpleColumnsComponent.prototype.getCheckAllState = /**
     * 0: 1个也没有选中； 1: 全部选中； 2: 有选中的也有没选中的
     * @private
     * @return {?}
     */
    function () {
        if (this.fields.length === 0) {
            return 0;
        }
        else {
            if (this.fields.length === this.columns[0].length) {
                return 1;
            }
            return 2;
        }
    };
    /**
     * @param {?} $event
     * @param {?} ct
     * @param {?} txt
     * @return {?}
     */
    SimpleColumnsComponent.prototype.onCheckAllClick = /**
     * @param {?} $event
     * @param {?} ct
     * @param {?} txt
     * @return {?}
     */
    function ($event, ct, txt) {
        /** @type {?} */
        var state = this.getCheckAllState();
        if (state === 0 || state === 2) {
            // 执行全选
            this.columns[0].forEach((/**
             * @param {?} n
             * @return {?}
             */
            function (n) { return n.visible = true; }));
            this.fields = this.columns[0].map((/**
             * @param {?} n
             * @return {?}
             */
            function (n) { return n.field; }));
        }
        else {
            // 取消全选
            this.columns[0].forEach((/**
             * @param {?} n
             * @return {?}
             */
            function (n) { return n.visible = false; }));
            this.fields = [];
        }
        this.onFieldsChanged();
        this.filterColumns(ct, txt);
    };
    /**
     * @param {?=} $event
     * @return {?}
     */
    SimpleColumnsComponent.prototype.close = /**
     * @param {?=} $event
     * @return {?}
     */
    function ($event) {
        if ($event) {
            $event.stopPropagation();
        }
        this.closed.emit();
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    SimpleColumnsComponent.prototype.showAdvanced = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        this.close($event);
        this.advanced.emit();
    };
    /**
     * @param {?=} $event
     * @return {?}
     */
    SimpleColumnsComponent.prototype.onSettingReset = /**
     * @param {?=} $event
     * @return {?}
     */
    function ($event) {
        var _this = this;
        if ($event) {
            $event.stopPropagation();
        }
        /** @type {?} */
        var restorDefaultText = this.localeSer.getValue('datagrid.settings.restoreDefaultSettingsText') || '确认要恢复默认设置吗？';
        this.messager.confirm(restorDefaultText).subscribe((/**
         * @param {?} t
         * @return {?}
         */
        function (t) {
            if (t) {
                _this.reset2Default();
            }
        }));
    };
    /**
     * @private
     * @return {?}
     */
    SimpleColumnsComponent.prototype.reset2Default = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var options = this.gridInstance.dfs['_state'].initialOptions;
        if (options.columnFields && options.columnFields.length) {
            /** @type {?} */
            var newColumns_1 = [];
            options.columnFields.forEach((/**
             * @param {?} c
             * @return {?}
             */
            function (c) {
                if (c.field !== '_datagrid-setting-control_') {
                    /** @type {?} */
                    var col = _this.gridInstance.columns[0].find((/**
                     * @param {?} n
                     * @return {?}
                     */
                    function (n) { return n.field === c.field; }));
                    if (col) {
                        col.visible = c.visible;
                        col.halign = c.halign;
                        col.align = c.align;
                        col.width = c.width;
                        col.footer = tslib_1.__assign({}, c.footer || {});
                        col.groupFooter = tslib_1.__assign({}, c.groupFooter || {});
                        newColumns_1.push(col);
                    }
                }
            }));
            this.columns = [newColumns_1];
        }
        this.allCols = this.columns[0];
        this.fields = this.getViewColumns();
        this.settings.viewColumns = this.fields;
        this.settings.columnFormat = tslib_1.__spread(this.columns[0]);
        if (options.sort) {
            /** @type {?} */
            var sortOrders = options.sort.sortOrder ? options.sort.sortOrder.split(',') : [];
            /** @type {?} */
            var sortNames = options.sort.sortName ? options.sort.sortName.split(',') : [];
            this.settings.sortInfo = { sortName: sortNames, sortOrder: sortOrders };
        }
        else {
            this.settings.sortInfo = {};
        }
        this.settings.groupFields = options.groupField ? options.groupField.split(',') : [];
    };
    /**
     * @param {?=} $event
     * @return {?}
     */
    SimpleColumnsComponent.prototype.onSubmitClick = /**
     * @param {?=} $event
     * @return {?}
     */
    function ($event) {
        if ($event) {
            $event.stopPropagation();
        }
        this.settings.viewColumns = this.fields;
        this.settings.columnFormat = tslib_1.__spread(this.columns[0]);
        if (!this.settings.sortInfo) {
            this.settings.sortInfo = {
                sortName: this.gridInstance.sortName ? this.gridInstance.sortName.split(',') : [],
                sortOrder: this.gridInstance.sortOrder ? this.gridInstance.sortOrder.split(',') : []
            };
        }
        if (this.settings.groupFields === undefined) {
            this.settings.groupFields = this.gridInstance.groupField ? this.gridInstance.groupField.split(',') : [];
        }
        this.submit.emit({ content: this.settings, close: this.close.bind(this), target: this.btnOk.nativeElement });
    };
    SimpleColumnsComponent.decorators = [
        { type: Component, args: [{
                    selector: 'datagrid-settings-simple',
                    template: "<div class=\"d-flex flex-column f-datagrid-settings-simple\" tabindex=\"0\" #colcontainer\r\n    [style.top.px]=\"top\" [style.height]=\"'calc(100vh - ' +  top + 'px)'\">\r\n\r\n    <div class=\"header d-flex\">\r\n        <div class=\"f-utils-fill title\"> {{ 'datagrid.settings.simple.title' | locale }} </div>\r\n        <div class=\"closed\" (click)=\"close($event)\">\r\n            <span class=\"f-icon f-icon-close\" style=\"font-size: 20px;\"></span>\r\n        </div>\r\n    </div>\r\n    <div class=\"f-utils-fill content d-flex flex-column\">\r\n        <div>\r\n            <h5 class=\"count\" [innerHTML]=\"'datagrid.settings.simple.count' | locale | replaceX: fields.length\">\r\n            </h5>\r\n\r\n            <div class=\"tip\">\r\n                {{ 'datagrid.settings.simple.tip' | locale }}\r\n            </div>\r\n\r\n            <input #inputWord type=\"text\" class=\"form-control\" placeholder=\"{{'datagrid.settings.searchPlaceholder' | locale}}\"\r\n                (input)=\"searchColumn($event, ct.selectedValues)\" style=\"height: 30px;margin-bottom: 16px;\">\r\n            <div style=\"padding: 5px 0px;background-color: #fff;\">\r\n                <button class=\"btn btn-link \" \r\n                    style=\"height: 32px;padding-left: 4px;\" (click)=\"onCheckAllClick($event, ct.selectedValues, inputWord.value)\">\r\n                    <span class=\"f-icon f-icon-minus-outline checkall\" [class.f-icon-checkbox]=\"!fields.length\"\r\n                        [class.f-icon-checkbox-checked]=\"fields.length === columns[0].length\"></span>\r\n                    {{ 'datagrid.settings.checkall' | locale }}\r\n                </button>\r\n\r\n                <farris-combo-list #ct \r\n                [data]=\"seartTypes\" \r\n                [editable]=\"false\"\r\n                [ngModel]=\"'all'\" [idField]=\"'value'\"\r\n                [textField]=\"'title'\"\r\n                [enableClear]=\"false\"\r\n                (valueChange)=\"filterColumns(ct.selectedValues, inputWord.value)\"\r\n                style=\"float: right;\" [style.width.px]=\"isEn ? 120: 80\"\r\n                class=\"format-setting-searchitem\"></farris-combo-list>\r\n            </div>\r\n        </div>\r\n        <div style=\"overflow: auto;padding-right:3px\" class=\"f-utils-fill\">\r\n            <ul class=\"list-group list-group-flush\" #ul column-visible-drag (drop)=\"onFieldDrop($event, ct.selectedValues, inputWord.value, ul)\">\r\n                <li class=\"list-group-item list-group-item-action smooth-dnd-draggable-wrapper format-setting-column\" *ngFor=\"let col of allCols\" \r\n                    (click)=\"onDblClick($event, col)\">\r\n                    <span class=\"f-icon checkbox\"\r\n                        [class.f-icon-checkbox-checked]=\"fields.indexOf(col.field) > -1\"\r\n                        [class.f-icon-checkbox]=\"fields.indexOf(col.field) == -1\"\r\n                        (click)=\"onCheckboxClick($event, col)\"></span>\r\n                    {{ col.title}} \r\n\r\n                    <span class=\"f-icon f-icon-drag-vertical field-drag-handle\"></span>\r\n                </li>\r\n            </ul>\r\n        </div>\r\n\r\n    </div>\r\n    <div class=\"footer\">\r\n\r\n        <div class=\"buttons\">\r\n            <button class=\"btn btn-link advanced\" (click)=\"showAdvanced($event)\">{{ 'datagrid.settings.advancedMode' | locale: '\u9AD8\u7EA7\u6A21\u5F0F' }}</button>\r\n            <button class=\"btn btn-link reset\" (click)=\"onSettingReset($event)\">{{ 'datagrid.settings.reset' | locale: '\u6062\u590D\u9ED8\u8BA4' }}</button>\r\n            <button class=\"btn btn-secondary cancel\"\r\n                (click)=\"close($event)\">{{'datagrid.settings.cancel' | locale}}</button>\r\n            <button class=\"btn btn-primary ok\" #okbtn\r\n                (click)=\"onSubmitClick($event)\">{{'datagrid.settings.ok' | locale}}</button>\r\n        </div>\r\n\r\n    </div>\r\n\r\n</div>",
                    encapsulation: ViewEncapsulation.None,
                    styles: ["\n        .f-datagrid-settings-simple {\n            position: absolute;\n            top: 100px;\n            right: 0px;\n            z-index: 10000;\n            height: calc(100vh - 100px);\n            width: 468px;\n            background: #FFFFFF;\n            border: 1px solid #E6E9F0;\n            border-radius: 10px 0 0 10px;\n        }\n\n        .f-datagrid-settings-simple:focus {\n            outline: none;\n        }\n\n        .f-datagrid-settings-simple .header {\n            height: 50px;\n            line-height: 50px;\n            border-bottom: 1px solid #EAEDF3;\n            font-size: 18px;\n            color: #2D2F33;\n        }\n        .f-datagrid-settings-simple .header .title {\n            padding-left: 21px;\n            font-weight:bold;\n        }\n        .f-datagrid-settings-simple .header .closed\n        {\n            width: 50px;\n            text-align: center;\n            cursor: pointer;\n        }\n\n        .f-datagrid-settings-simple .content .count{font-size: 14px; color: #2D2F33;}\n        .f-datagrid-settings-simple .content .tip {\n            background: #F7FAFF; border-radius: 6px;line-height: 42px; padding-left: 10px; margin-bottom: 10px;\n        }\n\n        .f-datagrid-settings-simple .content .checkall{\n            margin-left: 5px;\n            font-size: 14px;\n            color: #0089FF;\n            margin-right: 5px;\n        }\n\n        .f-datagrid-settings-simple .content {\n            margin: 20px 24px;\n        }\n\n        .f-datagrid-settings-simple .footer {\n            height: 32px;\n            margin: 20px;\n        }\n\n\n        .smooth-dnd-ghost.format-setting-column {\n            z-index: 999999999 !important;\n            padding: 4px 8px;\n            background-color: #e7f1ff;\n            border: 1px solid #2a87ff;\n        }\n        .smooth-dnd-ghost.format-setting-column .field-drag-handle {\n            float: right;\n        }\n        .f-datagrid-settings-simple  .smooth-dnd-drop-preview-constant-class {\n            background-color: #dee1ea;\n        }\n        .f-datagrid-settings-simple .format-setting-column {\n            background: #fafbfd;\n            padding: 4px 8px;\n            margin-bottom: 5px;\n            border: 1px solid #f2f4f8;\n            border-radius: 5px;\n            color: #000;\n        }\n        .f-datagrid-settings-simple .format-setting-column.active {\n            border: 1px solid #2a87ff;\n            background: #e7f1ff;\n            border-top: 1px solid #2a87ff!important;\n            border-bottom: 1px solid #2a87ff!important;\n        }\n\n        .format-setting-column .checkbox{\n            font-size: 14px;\n            color: #0089FF;\n            margin-right: 5px;\n            cursor:pointer;\n        }\n\n        .f-datagrid-settings-simple .format-setting-column .field-drag-handle{\n            float: right;\n            cursor: grab;\n        }\n\n        .f-datagrid-settings-simple .format-setting-column .field-drag-handle:active{\n            cursor: grabbing;\n        }\n\n        .f-datagrid-settings-simple .list-group-flush:last-child .list-group-item:last-child {\n            border-bottom: 1px solid #f2f4f8;\n        }\n\n        .f-datagrid-settings-simple .list-group-flush:first-child .list-group-item:first-child {\n            border-top: 1px solid #f2f4f8;\n        }\n\n        .f-datagrid-settings-simple .buttons\n        {\n            text-align: right;\n        }\n\n        .f-datagrid-settings-simple .buttons button {\n            width: 66px;height: 32px;\n        }\n        .f-datagrid-settings-simple .buttons .cancel {\n            margin-left: 12px;margin-right: 8px;\n        }\n        .f-datagrid-settings-simple .buttons .ok {\n            width: 100px;\n        }\n\n        .f-datagrid-settings-simple .buttons .advanced{ width: 66px;height: 32px; float: left; }\n\n        .f-datagrid-settings-simple-host {\n            width: 100%;\n            height: 100%;\n            position: absolute;\n            top: 0;\n            left: 0;\n            z-index: 1000;\n        }\n\n        .f-datagrid-settings-simple .visible-cols-count {\n            font-size: 18px;\n            font-weight: 500;\n        }\n        "]
                }] }
    ];
    /** @nocollapse */
    SimpleColumnsComponent.ctorParameters = function () { return [
        { type: Injector },
        { type: ChangeDetectorRef }
    ]; };
    SimpleColumnsComponent.propDecorators = {
        seartTypes: [{ type: Input }],
        columns: [{ type: Input }],
        fields: [{ type: Input }],
        clickToHidden: [{ type: Input }],
        top: [{ type: Input }],
        cls: [{ type: HostBinding, args: ['class',] }],
        container: [{ type: ViewChild, args: ['colcontainer',] }],
        btnOk: [{ type: ViewChild, args: ['okbtn',] }],
        searchInput: [{ type: ViewChild, args: ['inputWord',] }],
        formatChange: [{ type: Output }],
        fieldsChange: [{ type: Output }],
        submit: [{ type: Output }],
        closed: [{ type: Output }],
        advanced: [{ type: Output }],
        onClick: [{ type: HostListener, args: ['click', ['$event'],] }],
        onEscKeydown: [{ type: HostListener, args: ['window:keydown.esc', ['$event'],] }],
        onEnterKeydown: [{ type: HostListener, args: ['window:keydown.enter', ['$event'],] }],
        onControlGKeydown: [{ type: HostListener, args: ['keydown', ['$event'],] }]
    };
    return SimpleColumnsComponent;
}());
export { SimpleColumnsComponent };
if (false) {
    /** @type {?} */
    SimpleColumnsComponent.prototype.seartTypes;
    /**
     * 所有列
     * @type {?}
     */
    SimpleColumnsComponent.prototype.columns;
    /**
     * 显示的列
     * @type {?}
     */
    SimpleColumnsComponent.prototype.fields;
    /**
     * 点击侧边样式以外的区域隐藏面板。默认为 true
     * @type {?}
     */
    SimpleColumnsComponent.prototype.clickToHidden;
    /** @type {?} */
    SimpleColumnsComponent.prototype.top;
    /** @type {?} */
    SimpleColumnsComponent.prototype.cls;
    /** @type {?} */
    SimpleColumnsComponent.prototype.container;
    /** @type {?} */
    SimpleColumnsComponent.prototype.btnOk;
    /** @type {?} */
    SimpleColumnsComponent.prototype.searchInput;
    /** @type {?} */
    SimpleColumnsComponent.prototype.formatChange;
    /** @type {?} */
    SimpleColumnsComponent.prototype.fieldsChange;
    /** @type {?} */
    SimpleColumnsComponent.prototype.submit;
    /** @type {?} */
    SimpleColumnsComponent.prototype.closed;
    /** @type {?} */
    SimpleColumnsComponent.prototype.advanced;
    /** @type {?} */
    SimpleColumnsComponent.prototype.allCols;
    /** @type {?} */
    SimpleColumnsComponent.prototype.columnType;
    /** @type {?} */
    SimpleColumnsComponent.prototype.gridInstance;
    /** @type {?} */
    SimpleColumnsComponent.prototype.localeSer;
    /** @type {?} */
    SimpleColumnsComponent.prototype.isEn;
    /**
     * @type {?}
     * @private
     */
    SimpleColumnsComponent.prototype.messager;
    /**
     * @type {?}
     * @private
     */
    SimpleColumnsComponent.prototype.settings;
    /**
     * @type {?}
     * @private
     */
    SimpleColumnsComponent.prototype.injector;
    /** @type {?} */
    SimpleColumnsComponent.prototype.cd;
}
