// @farris/ui-personnel-selector/esm2015/farris-ui-personnel-selector.js
/**
 * @fileoverview added by tsickle
 * Generated from: farris-ui-personnel-selector.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Generated bundle index. Do not edit.
 */
export { DatagridPersonnelSelectorComponent, PersonnelSelectorDataGridEditorProvider, PersonnelSelectorComponent, PersonnelSelectorModule, SelectionService, CandidateComponent, FavoritesComponent, GroupComponent, UnSelectDatagridComponent, UserGroupUnSelectComponent, RecentContactsComponent, ApiLocalService, GROUP_SERVICE, FavoritesTool, GroupTool, FilterOrg, DataLoadingService, FilterOrgTool } from './public-api';
export { QuickSelectComponent as ɵf } from './lib/quick-select-page/qsp.component';
export { quickInputDatagridComponent as ɵa } from './lib/quickInputDatagrid.component';
export { SearchResultComponent as ɵe } from './lib/search-result-page/srp.component';
export { OrgPersonnelComponent as ɵd } from './lib/tab-org-personnel/op.component';
export { unSelectDatagridRoleComponent as ɵb } from './lib/unSelectDatagridRoleComponent';
export { unSelectDatagridOrgComponent as ɵc } from './lib/unselectdatagridorg.component';
