// @gsp-cmp/chgdr/esm2015/lib/service/chgdr-metadata.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Injector } from '@angular/core';
import { HttpService, SessionService } from '@ecp-caf/caf-common';
import { GSPMetadataRTService } from '@gsp-lcm/metadatart-selector';
import { of } from 'rxjs';
import { map } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@ecp-caf/caf-common";
import * as i2 from "@gsp-lcm/metadatart-selector";
export class ChgdrMetadataService {
    /**
     * @param {?} http
     * @param {?} sessionService
     * @param {?} metadataService
     * @param {?} injector
     */
    constructor(http, sessionService, metadataService, injector) {
        this.http = http;
        this.sessionService = sessionService;
        this.metadataService = metadataService;
        this.injector = injector;
        this.beMetadataCache = new Map();
    }
    /**
     * 获取BE元数据
     * @param {?} beId
     * @return {?}
     */
    getGSPBusinessEntity(beId) {
        if (this.beMetadataCache.has(beId)) {
            /** @type {?} */
            let be = this.beMetadataCache.get(beId);
            return of(be);
        }
        else {
            return this.metadataService.GetMetadataRT(beId).pipe(map((/**
             * @param {?} metadata
             * @return {?}
             */
            metadata => {
                /** @type {?} */
                const be = JSON.parse(metadata.content);
                this.beMetadataCache.set(beId, be);
                return be;
            })));
        }
    }
}
ChgdrMetadataService.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
/** @nocollapse */
ChgdrMetadataService.ctorParameters = () => [
    { type: HttpService },
    { type: SessionService },
    { type: GSPMetadataRTService },
    { type: Injector }
];
/** @nocollapse */ ChgdrMetadataService.ngInjectableDef = i0.defineInjectable({ factory: function ChgdrMetadataService_Factory() { return new ChgdrMetadataService(i0.inject(i1.HttpService), i0.inject(i1.SessionService), i0.inject(i2.GSPMetadataRTService), i0.inject(i0.INJECTOR)); }, token: ChgdrMetadataService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    ChgdrMetadataService.prototype.beMetadataCache;
    /**
     * @type {?}
     * @private
     */
    ChgdrMetadataService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    ChgdrMetadataService.prototype.sessionService;
    /**
     * @type {?}
     * @private
     */
    ChgdrMetadataService.prototype.metadataService;
    /**
     * @type {?}
     * @private
     */
    ChgdrMetadataService.prototype.injector;
}
