// @farris/ui-common/esm5/date/lib/date-helper.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * @Author: 疯狂秀才(Lucas Huang)
 * @Date: 2019-09-02 17:55:57
 * @LastEditors: 疯狂秀才(Lucas Huang)
 * @LastEditTime: 2019-10-23 16:57:28
 * @QQ: 1055818239
 * @Version: v0.0.1
 */
import { format, isValid, parseISO } from 'date-fns';
import { Injectable } from '@angular/core';
import { MyDateConverter } from './myDateConverter';
import * as i0 from "@angular/core";
var DateTimeHelperService = /** @class */ (function () {
    function DateTimeHelperService() {
        this.$D = null;
        this.$D = new MyDateConverter();
    }
    /**
     * @param value 要转换格式的日期
     * @param fmt 格式化字符串
     *
     * 更多的格式化请参考
     * [点我点我](https://date-fns.org/v1.30.1/docs/format)
     */
    /**
     * @param {?} value 要转换格式的日期
     * @param {?=} fmt 格式化字符串
     *
     * 更多的格式化请参考
     * [点我点我](https://date-fns.org/v1.30.1/docs/format)
     * @return {?}
     */
    DateTimeHelperService.prototype.formatTo = /**
     * @param {?} value 要转换格式的日期
     * @param {?=} fmt 格式化字符串
     *
     * 更多的格式化请参考
     * [点我点我](https://date-fns.org/v1.30.1/docs/format)
     * @return {?}
     */
    function (value, fmt) {
        if (fmt === void 0) { fmt = 'yyyy-MM-dd'; }
        if (!value) {
            return '';
        }
        if (typeof value === 'string' && value.indexOf('0001') > -1) {
            return '';
        }
        if (value instanceof Date) {
            return format(value, fmt);
        }
        /** @type {?} */
        var _d = parseISO(value);
        if (_d == 'Invalid Date') {
            _d = this.createDate(value) || new Date(value);
        }
        // const d = parseISO(value);
        if (isValid(_d)) {
            /** @type {?} */
            var d = parseISO(format(_d, 'yyyy-MM-dd HH:mm:ss'));
            return format(d, fmt);
        }
        else {
            throw new Error('给定的值不能转换为日期');
        }
    };
    /**
     * @private
     * @param {?} val
     * @return {?}
     */
    DateTimeHelperService.prototype.createDate = /**
     * @private
     * @param {?} val
     * @return {?}
     */
    function (val) {
        return this.$D.create(val);
    };
    DateTimeHelperService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    DateTimeHelperService.ctorParameters = function () { return []; };
    /** @nocollapse */ DateTimeHelperService.ngInjectableDef = i0.defineInjectable({ factory: function DateTimeHelperService_Factory() { return new DateTimeHelperService(); }, token: DateTimeHelperService, providedIn: "root" });
    return DateTimeHelperService;
}());
export { DateTimeHelperService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    DateTimeHelperService.prototype.$D;
}
