// @gsp-cmp/chgdr/esm5/lib/service/chgdr-config.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Injector } from '@angular/core';
import { HttpService, SessionService } from '@ecp-caf/caf-common';
import { tap } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@ecp-caf/caf-common";
/** @type {?} */
var CHGDR_CONFIG_ROUTER = '/api/runtime/bcc/v1.0/chgdr/config';
/** @type {?} */
var CHGDR_CONFIG_GET_BY_BIZOBJID_AND_MODELID = CHGDR_CONFIG_ROUTER + "/{bizObjId}/{modelId}";
/** @type {?} */
var ServerIP = '/';
/** @type {?} */
var chgdrUrl = ServerIP + "api/runtime/chgdr/v1.0";
var ChgdrConfigService = /** @class */ (function () {
    function ChgdrConfigService(http, sessionService, injector) {
        this.http = http;
        this.sessionService = sessionService;
        this.injector = injector;
    }
    /**
     * 根据业务实体ID获取所有的配置信息
     */
    /**
     * 根据业务实体ID获取所有的配置信息
     * @param {?} beId
     * @return {?}
     */
    ChgdrConfigService.prototype.getBeChgdrConfig = /**
     * 根据业务实体ID获取所有的配置信息
     * @param {?} beId
     * @return {?}
     */
    function (beId) {
        /** @type {?} */
        var queryUrl = chgdrUrl + "/config?beId=" + beId;
        return ((/** @type {?} */ ((/** @type {?} */ (this.http.get(queryUrl)))))).pipe(tap((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            /** @type {?} */
            var configEntities = data || [];
            configEntities.forEach((/**
             * @param {?} entity
             * @return {?}
             */
            function (entity) {
                //处理字段顺序号
                /** @type {?} */
                var fields = entity.fields || [];
                //赋予老数据默认顺序号
                fields.forEach((/**
                 * @param {?} field
                 * @param {?} index
                 * @return {?}
                 */
                function (field, index) {
                    if (field.showIndex === undefined || field.showIndex === null) {
                        field.showIndex = index;
                    }
                }));
                //按照顺序号排序
                entity.fields = fields.sort((/**
                 * @param {?} a
                 * @param {?} b
                 * @return {?}
                 */
                function (a, b) { return a.showIndex - b.showIndex; }));
                //根据排序结果重置顺序号
                fields.forEach((/**
                 * @param {?} field
                 * @param {?} index
                 * @return {?}
                 */
                function (field, index) {
                    field.showIndex = index;
                }));
            }));
        })));
    };
    ChgdrConfigService.decorators = [
        { type: Injectable, args: [{ providedIn: 'root' },] }
    ];
    /** @nocollapse */
    ChgdrConfigService.ctorParameters = function () { return [
        { type: HttpService },
        { type: SessionService },
        { type: Injector }
    ]; };
    /** @nocollapse */ ChgdrConfigService.ngInjectableDef = i0.defineInjectable({ factory: function ChgdrConfigService_Factory() { return new ChgdrConfigService(i0.inject(i1.HttpService), i0.inject(i1.SessionService), i0.inject(i0.INJECTOR)); }, token: ChgdrConfigService, providedIn: "root" });
    return ChgdrConfigService;
}());
export { ChgdrConfigService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ChgdrConfigService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    ChgdrConfigService.prototype.sessionService;
    /**
     * @type {?}
     * @private
     */
    ChgdrConfigService.prototype.injector;
}
