// @gsp-cmp/chgdr/esm5/lib/service/chgdr-inner.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Injector } from '@angular/core';
import { HttpService, SessionService } from '@ecp-caf/caf-common';
import { GSPMetadataRTService } from '@gsp-lcm/metadatart-selector';
import { forkJoin, of } from 'rxjs';
import { map } from 'rxjs/operators';
import { ChgdrConfigHandler } from './chgdr-config-handler';
import { ChgdrConfigService } from './chgdr-config.service';
import { ChgdrMetadataService } from './chgdr-metadata.service';
import * as i0 from "@angular/core";
import * as i1 from "@ecp-caf/caf-common";
import * as i2 from "@gsp-lcm/metadatart-selector";
import * as i3 from "./chgdr-metadata.service";
import * as i4 from "./chgdr-config.service";
var ChgdrInnerService = /** @class */ (function () {
    function ChgdrInnerService(http, sessionService, metadataService, chgdrMetadataService, chgdrConfigService, injector) {
        this.http = http;
        this.sessionService = sessionService;
        this.metadataService = metadataService;
        this.chgdrMetadataService = chgdrMetadataService;
        this.chgdrConfigService = chgdrConfigService;
        this.injector = injector;
        this.beMetadataCache = new Map();
    }
    /** 获取BE元数据 */
    /**
     * 获取BE元数据
     * @param {?} beId
     * @return {?}
     */
    ChgdrInnerService.prototype.getChgdrConfigHandler = /**
     * 获取BE元数据
     * @param {?} beId
     * @return {?}
     */
    function (beId) {
        var _this = this;
        if (this.beMetadataCache.has(beId)) {
            /** @type {?} */
            var be = this.beMetadataCache.get(beId);
            return of(be);
        }
        else {
            /** @type {?} */
            var requests = [];
            requests.push(this.chgdrMetadataService.getGSPBusinessEntity(beId));
            requests.push(this.chgdrConfigService.getBeChgdrConfig(beId));
            return forkJoin(requests).pipe(map((/**
             * @param {?} data
             * @return {?}
             */
            function (data) {
                /** @type {?} */
                var be = data[0];
                /** @type {?} */
                var entityConfigs = data[1] || [];
                /** @type {?} */
                var handler = new ChgdrConfigHandler(be, entityConfigs);
                _this.beMetadataCache.set(beId, handler);
                return handler;
            })));
        }
    };
    ChgdrInnerService.decorators = [
        { type: Injectable, args: [{ providedIn: 'root' },] }
    ];
    /** @nocollapse */
    ChgdrInnerService.ctorParameters = function () { return [
        { type: HttpService },
        { type: SessionService },
        { type: GSPMetadataRTService },
        { type: ChgdrMetadataService },
        { type: ChgdrConfigService },
        { type: Injector }
    ]; };
    /** @nocollapse */ ChgdrInnerService.ngInjectableDef = i0.defineInjectable({ factory: function ChgdrInnerService_Factory() { return new ChgdrInnerService(i0.inject(i1.HttpService), i0.inject(i1.SessionService), i0.inject(i2.GSPMetadataRTService), i0.inject(i3.ChgdrMetadataService), i0.inject(i4.ChgdrConfigService), i0.inject(i0.INJECTOR)); }, token: ChgdrInnerService, providedIn: "root" });
    return ChgdrInnerService;
}());
export { ChgdrInnerService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ChgdrInnerService.prototype.beMetadataCache;
    /**
     * @type {?}
     * @private
     */
    ChgdrInnerService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    ChgdrInnerService.prototype.sessionService;
    /**
     * @type {?}
     * @private
     */
    ChgdrInnerService.prototype.metadataService;
    /**
     * @type {?}
     * @private
     */
    ChgdrInnerService.prototype.chgdrMetadataService;
    /**
     * @type {?}
     * @private
     */
    ChgdrInnerService.prototype.chgdrConfigService;
    /**
     * @type {?}
     * @private
     */
    ChgdrInnerService.prototype.injector;
}
