// @farris/ui-filter/esm2015/lib/entity/conditionvalue/datevalue.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ControlType } from '../controltype';
import { EventEmitter } from '@angular/core';
export class DateValue {
    /**
     * @param {?=} value
     */
    constructor(value = { value: '' }) {
        this.clear$ = new EventEmitter();
        this.value = value.value;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setOriginalValue(value) {
        this.value = value.formatted;
    }
    /**
     * @return {?}
     */
    getOriginalValue() {
        return this.value;
    }
    /**
     * @return {?}
     */
    getControlType() {
        return ControlType.Date;
    }
    /**
     * @return {?}
     */
    isEmpty() {
        return !this.value;
    }
    /**
     * @return {?}
     */
    clearValue() {
        this.value = undefined;
        this.clear$.emit();
    }
    /**
     * @return {?}
     */
    getValueText() {
        return this.value;
    }
}
if (false) {
    /** @type {?} */
    DateValue.prototype.value;
    /** @type {?} */
    DateValue.prototype.clear$;
}
