// @gsp-cmp/chgdr/esm5/lib/chgdr-list-dialog-content/chgdr-list-dialog-content.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Injector, Input } from '@angular/core';
var ChgdrListDialogContentComponent = /** @class */ (function () {
    function ChgdrListDialogContentComponent(injector) {
        this.injector = injector;
    }
    /**
     * @return {?}
     */
    ChgdrListDialogContentComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    ChgdrListDialogContentComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lib-chgdr-list-dialog-content',
                    template: "<chgdr-list [entityId]=\"entityId\" [dataId]=\"dataId\" style=\"padding: 0.75rem 0.875rem;height:100%\"></chgdr-list>",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    ChgdrListDialogContentComponent.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    ChgdrListDialogContentComponent.propDecorators = {
        entityId: [{ type: Input }],
        dataId: [{ type: Input }]
    };
    return ChgdrListDialogContentComponent;
}());
export { ChgdrListDialogContentComponent };
if (false) {
    /** @type {?} */
    ChgdrListDialogContentComponent.prototype.entityId;
    /** @type {?} */
    ChgdrListDialogContentComponent.prototype.dataId;
    /**
     * @type {?}
     * @private
     */
    ChgdrListDialogContentComponent.prototype.injector;
}
