// @farris/component-querycondition/esm2015/lib/directives/clear-fields.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/directives/clear-fields.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, Optional, Self, Injector } from '@angular/core';
import { NgControl } from '@angular/forms';
import { QueryConditionComponent } from '../querycondition.component';
export class ClearFieldsDirective {
    /**
     * @param {?} ngCtrl
     * @param {?} qcc
     * @param {?} inject
     */
    constructor(ngCtrl, qcc, inject) {
        this.ngCtrl = ngCtrl;
        this.qcc = qcc;
        this.inject = inject;
        this.clearFields = '';
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (this.clearFields && this.ngCtrl) {
            //console.log(this.ngCtrl, this.clearFields, this.qcc);
            //console.log(this.qcc.selectedFieldConditions);
            this.ngCtrl.valueChanges.subscribe((/**
             * @param {?} v
             * @return {?}
             */
            (v) => {
                this.qcc.fieldCodeOfClearFields = [];
                this.clearFields.split(',').forEach((/**
                 * @param {?} n
                 * @return {?}
                 */
                n => {
                    /** @type {?} */
                    const conditionItem = this.qcc.selectedFieldConditions.find((/**
                     * @param {?} c
                     * @return {?}
                     */
                    c => c.fieldCode === n));
                    if (conditionItem) {
                        this.qcc.fieldCodeOfClearFields.push(n);
                        conditionItem.value.clearValue();
                    }
                }));
            }));
        }
    }
}
ClearFieldsDirective.decorators = [
    { type: Directive, args: [{
                selector: '[clear-fields]'
            },] }
];
/** @nocollapse */
ClearFieldsDirective.ctorParameters = () => [
    { type: NgControl, decorators: [{ type: Optional }, { type: Self }] },
    { type: QueryConditionComponent },
    { type: Injector }
];
ClearFieldsDirective.propDecorators = {
    clearFields: [{ type: Input, args: ['clear-fields',] }]
};
if (false) {
    /** @type {?} */
    ClearFieldsDirective.prototype.clearFields;
    /**
     * @type {?}
     * @private
     */
    ClearFieldsDirective.prototype.ngCtrl;
    /**
     * @type {?}
     * @private
     */
    ClearFieldsDirective.prototype.qcc;
    /**
     * @type {?}
     * @private
     */
    ClearFieldsDirective.prototype.inject;
}
