// @farris/devkit/esm2015/lib/core/types.js
import { InjectionToken } from '@angular/core';
// export type ClassType = new (...args: any[]) => any;
export const BigNumberType = 'BigNumber';
export var DataChangeType;
(function (DataChangeType) {
    DataChangeType[DataChangeType["Add"] = 0] = "Add";
    DataChangeType[DataChangeType["Delete"] = 1] = "Delete";
})(DataChangeType || (DataChangeType = {}));
export var HttpMethod;
(function (HttpMethod) {
    HttpMethod["GET"] = "GET";
    HttpMethod["POST"] = "POST";
    HttpMethod["PUT"] = "PUT";
    HttpMethod["DELETE"] = "DELETE";
})(HttpMethod || (HttpMethod = {}));
/**
 * 后端消息
 */
// tslint:disable-next-line: no-namespace
export var BackEndMessage;
(function (BackEndMessage) {
    /**
     * 消息级别
     */
    let Level;
    (function (Level) {
        Level["Error"] = "Error";
        Level["Info"] = "Info";
        Level["Warning"] = "Warning";
    })(Level = BackEndMessage.Level || (BackEndMessage.Level = {}));
    /**
     * 消息
     * @description 接口执行成功，但有info/warning级别的消息
     */
    class Message {
        constructor(bizMessages, context) {
            this.bizMessages = bizMessages;
            this.context = context;
        }
    }
    BackEndMessage.Message = Message;
})(BackEndMessage || (BackEndMessage = {}));
export var RunMode;
(function (RunMode) {
    RunMode["compatible"] = "compatible";
    RunMode["highSpeed"] = "highSpeed";
})(RunMode || (RunMode = {}));
/**
 * 控制器运行模式，compatible | highSpeed
 */
export const DEVKIT_RUN_MODE = new InjectionToken('@farris/devkit_run_mode');
