// @farris/ui-filter/esm5/lib/entity/conditionvalue/monthrange.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ControlType } from '../controltype';
import { EventEmitter } from '@angular/core';
var MonthRangeValue = /** @class */ (function () {
    function MonthRangeValue(value) {
        if (value === void 0) { value = { startTime: '', endTime: '' }; }
        this.clear$ = new EventEmitter();
        this.startTime = value.startTime;
        this.endTime = value.endTime;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    MonthRangeValue.prototype.setOriginalValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (value.formatted) {
            this.startTime = value.formatted.split(value.delimiter)[0];
            this.endTime = value.formatted.split(value.delimiter)[1];
        }
        else {
            this.clearValue();
        }
    };
    /**
     * @return {?}
     */
    MonthRangeValue.prototype.getOriginalValue = /**
     * @return {?}
     */
    function () {
        return this.startTime + "~" + this.endTime;
    };
    /**
     * @return {?}
     */
    MonthRangeValue.prototype.getControlType = /**
     * @return {?}
     */
    function () {
        return ControlType.DateRange;
    };
    /**
     * @return {?}
     */
    MonthRangeValue.prototype.clearValue = /**
     * @return {?}
     */
    function () {
        this.startTime = undefined;
        this.endTime = undefined;
        this.clear$.emit();
    };
    /**
     * @return {?}
     */
    MonthRangeValue.prototype.isEmpty = /**
     * @return {?}
     */
    function () {
        return !this.startTime || !this.endTime;
    };
    /**
     * @return {?}
     */
    MonthRangeValue.prototype.getValueText = /**
     * @return {?}
     */
    function () {
        if (!this.startTime.length || !this.endTime.length) {
            return '';
        }
        else {
            return (this.startTime + '~' + this.endTime);
        }
    };
    return MonthRangeValue;
}());
export { MonthRangeValue };
if (false) {
    /** @type {?} */
    MonthRangeValue.prototype.startTime;
    /** @type {?} */
    MonthRangeValue.prototype.endTime;
    /** @type {?} */
    MonthRangeValue.prototype.clear$;
}
