// @farris/ui-filter/esm5/lib/entity/conditionvalue/yearvalue.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ControlType } from '../controltype';
import { EventEmitter } from '@angular/core';
var YearValue = /** @class */ (function () {
    function YearValue(value) {
        if (value === void 0) { value = { value: '' }; }
        this.clear$ = new EventEmitter();
        this.value = value.value;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    YearValue.prototype.setOriginalValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.value = value.formatted;
    };
    /**
     * @return {?}
     */
    YearValue.prototype.getOriginalValue = /**
     * @return {?}
     */
    function () {
        return this.value;
    };
    /**
     * @return {?}
     */
    YearValue.prototype.getControlType = /**
     * @return {?}
     */
    function () {
        return ControlType.Date;
    };
    /**
     * @return {?}
     */
    YearValue.prototype.isEmpty = /**
     * @return {?}
     */
    function () {
        return !this.value;
    };
    /**
     * @return {?}
     */
    YearValue.prototype.clearValue = /**
     * @return {?}
     */
    function () {
        this.value = undefined;
        this.clear$.emit();
    };
    /**
     * @return {?}
     */
    YearValue.prototype.getValueText = /**
     * @return {?}
     */
    function () {
        return this.value;
    };
    return YearValue;
}());
export { YearValue };
if (false) {
    /** @type {?} */
    YearValue.prototype.value;
    /** @type {?} */
    YearValue.prototype.clear$;
}
