// @edp-aif/common-api/esm2015/lib/billkit/BillSchemaFieldBase.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModelJsonHelper } from '../BpmnModelJsonHelper';
import { KitField } from './KitField';
import { DataType } from './DataType';
/**
 * @abstract
 */
export class BillSchemaFieldBase extends KitField {
    constructor() {
        super(...arguments);
        this.dataType = DataType.Unknow;
        this.extendProperties = {};
        this.length = 0;
        this.precision = 0;
        this.scale = 0;
    }
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.fieldType = null;
        this.dataType = DataType.Unknow;
        this.caption = null;
        this.extendProperties = {};
        this.length = 0;
        this.precision = 0;
        this.scale = 0;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "type", this.fieldType);
        if (this.dataType != DataType.String) {
            BpmnModelJsonHelper.SetAttributeValue(obj, "data", this.dataType);
        }
        BpmnModelJsonHelper.SetAttributeValue(obj, "cap", this.caption);
        BpmnModelJsonHelper.SetAttributeValue(obj, "props", Object.assign({}, this.extendProperties));
        BpmnModelJsonHelper.SetAttributeValue(obj, "length", this.length);
        BpmnModelJsonHelper.SetAttributeValue(obj, "precision", this.precision);
        BpmnModelJsonHelper.SetAttributeValue(obj, "scale", this.scale);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.fieldType = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "type");
        /** @type {?} */
        let data = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "data");
        if (data == null) {
            this.dataType = DataType.String;
        }
        else {
            this.dataType = DataType[data];
        }
        this.dataType = DataType[BpmnModelJsonHelper.GetAttributeValueAsString(obj, "data")];
        this.caption = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "cap");
        this.extendProperties = BpmnModelJsonHelper.GetAttributeValue(obj, "props");
        this.length = BpmnModelJsonHelper.GetAttributeValueAsNumber(obj, "length");
        this.precision = BpmnModelJsonHelper.GetAttributeValueAsNumber(obj, "precision");
        this.scale = BpmnModelJsonHelper.GetAttributeValueAsNumber(obj, "scale");
    }
}
if (false) {
    /** @type {?} */
    BillSchemaFieldBase.prototype.fieldType;
    /** @type {?} */
    BillSchemaFieldBase.prototype.dataType;
    /** @type {?} */
    BillSchemaFieldBase.prototype.caption;
    /** @type {?} */
    BillSchemaFieldBase.prototype.extendProperties;
    /** @type {?} */
    BillSchemaFieldBase.prototype.attachedField;
    /** @type {?} */
    BillSchemaFieldBase.prototype.length;
    /** @type {?} */
    BillSchemaFieldBase.prototype.precision;
    /** @type {?} */
    BillSchemaFieldBase.prototype.scale;
}
