// @edp-bif/common-api/esm2015/lib/entity/impl/AbstractBillCreationProcessEntity.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/AbstractBillCreationProcessEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModelJsonHelper } from '../../ref';
import { AbstractBillProcessEntity } from './AbstractBillProcessEntity';
/**
 * 单据生单流实体类。
 */
export class AbstractBillCreationProcessEntity extends AbstractBillProcessEntity {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.isSupportPull = false;
        this.isSupportPush = false;
        this.creationRuleId = null;
        this.creationProcessId = null;
        this.downBillVoProxyMetadataId = null;
        this.downBillVoProxyId = null;
        this.downBillVoProxyMetadataRootId = null;
        this.downBillVoProxyMetadataParId = null;
        this.downBillVoProxyMetadataIdPath = null;
        this.downBillVoProxyRootId = null;
        this.downBillVoProxyParentId = null;
        this.downBillVoProxyIdPath = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "isSupportPull", this.isSupportPull);
        BpmnModelJsonHelper.SetAttributeValue(obj, "isSupportPush", this.isSupportPush);
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleId", this.creationRuleId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationProcessId", this.creationProcessId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoProxyMetadataId", this.downBillVoProxyMetadataId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoProxyId", this.downBillVoProxyId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoProxyMetadataRootId", this.downBillVoProxyMetadataRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoProxyMetadataParId", this.downBillVoProxyMetadataParId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoProxyMetadataIdPath", this.downBillVoProxyMetadataIdPath);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoProxyRootId", this.downBillVoProxyRootId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoProxyParentId", this.downBillVoProxyParentId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillVoProxyIdPath", this.downBillVoProxyIdPath);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.isSupportPull = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isSupportPull");
        this.isSupportPush = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isSupportPush");
        this.creationRuleId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleId");
        this.creationProcessId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationProcessId");
        this.downBillVoProxyMetadataId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoProxyMetadataId");
        this.downBillVoProxyId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoProxyId");
        this.downBillVoProxyMetadataRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoProxyMetadataRootId");
        this.downBillVoProxyMetadataParId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoProxyMetadataParId");
        this.downBillVoProxyMetadataIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoProxyMetadataIdPath");
        this.downBillVoProxyRootId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoProxyRootId");
        this.downBillVoProxyParentId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoProxyParentId");
        this.downBillVoProxyIdPath = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillVoProxyIdPath");
    }
}
if (false) {
    /** @type {?} */
    AbstractBillCreationProcessEntity.prototype.downBillVoProxyMetadataRootId;
    /** @type {?} */
    AbstractBillCreationProcessEntity.prototype.downBillVoProxyMetadataParId;
    /** @type {?} */
    AbstractBillCreationProcessEntity.prototype.downBillVoProxyMetadataIdPath;
    /** @type {?} */
    AbstractBillCreationProcessEntity.prototype.downBillVoProxyRootId;
    /** @type {?} */
    AbstractBillCreationProcessEntity.prototype.downBillVoProxyParentId;
    /** @type {?} */
    AbstractBillCreationProcessEntity.prototype.downBillVoProxyIdPath;
    /** @type {?} */
    AbstractBillCreationProcessEntity.prototype.downBillVoProxyMetadataId;
    /** @type {?} */
    AbstractBillCreationProcessEntity.prototype.downBillVoProxyId;
    /** @type {?} */
    AbstractBillCreationProcessEntity.prototype.isSupportPull;
    /** @type {?} */
    AbstractBillCreationProcessEntity.prototype.isSupportPush;
    /** @type {?} */
    AbstractBillCreationProcessEntity.prototype.creationRuleId;
    /** @type {?} */
    AbstractBillCreationProcessEntity.prototype.creationProcessId;
}
