// @edp-bif/common-api/esm2015/lib/content/impl/RtBillCreationProcessContent.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/content/impl/RtBillCreationProcessContent.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BpmnModelJsonHelper } from '../../ref';
import { AbstractRtBillProcessContent } from './AbstractRtBillProcessContent';
import { RtBillCreationProcessEntity } from '../../entity/impl/RtBillCreationProcessEntity';
/**
 * 单据生单流内容类。
 */
export class RtBillCreationProcessContent extends AbstractRtBillProcessContent {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.formSettings = null;
        this.upBillTrackingSettings = null;
        this.downBillTrackingSettings = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "formSettings", this.formSettings);
        BpmnModelJsonHelper.SetAttributeValue(obj, "upBillTrackingSettings", this.upBillTrackingSettings);
        BpmnModelJsonHelper.SetAttributeValue(obj, "downBillTrackingSettings", this.downBillTrackingSettings);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.formSettings = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "formSettings");
        this.upBillTrackingSettings = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "upBillTrackingSettings");
        this.downBillTrackingSettings = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "downBillTrackingSettings");
    }
    /**
     * @protected
     * @return {?}
     */
    getEntityClass() {
        return RtBillCreationProcessEntity;
    }
}
if (false) {
    /** @type {?} */
    RtBillCreationProcessContent.prototype.formSettings;
    /** @type {?} */
    RtBillCreationProcessContent.prototype.upBillTrackingSettings;
    /** @type {?} */
    RtBillCreationProcessContent.prototype.downBillTrackingSettings;
}
