// @farris/ui-lookup/esm5/lib/lookup-displaytype.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * 收藏列字段名称
 * @type {?}
 */
export var FAVORITE_FIELD_NAME = '__favorite__';
/** @type {?} */
export var INPUT_GROUP_ICON = '<i class="f-icon f-icon-lookup"></i>';
/** @enum {string} */
var LookupGridDisplayType = {
    List: 'LIST',
    TreeList: 'TREELIST',
    NavList: 'NAVLIST',
    NavTreeList: 'NAVTREELIST',
};
export { LookupGridDisplayType };
/** @enum {string} */
var FavoriteIcon = {
    /** 已收藏 */
    yes: '<span class="f-icon f-icon-star f-lookup-favorite" style="line-height: 29px;"></span>',
    /** 未收藏 */
    no: '<span class="f-icon f-icon-star-outline f-lookup-favorite" style="line-height: 29px;"></span>',
    /** 移除收藏 */
    delete: '<span class="f-icon f-icon-apply-format f-lookup-unfavorite" style="line-height: 29px;"></span>',
    /** 移除已选记录 */
    remove: '<span class="f-icon f-icon-apply-format f-lookup-unfavorite" style="line-height: 29px;"></span>',
};
export { FavoriteIcon };
/** @enum {string} */
var FavoriteAction = {
    /** 添加收藏 */
    add: 'append item to favorite.',
    /** 移除收藏 */
    delete: 'remove favorite.',
};
export { FavoriteAction };
/**
 * @record
 */
export function ExtInfoFormatterParam() { }
if (false) {
    /** @type {?|undefined} */
    ExtInfoFormatterParam.prototype.bindingData;
    /** @type {?|undefined} */
    ExtInfoFormatterParam.prototype.instance;
    /* Skipping unhandled member: [key: string]: any;*/
}
/**
 * @record
 */
export function CustomDataFormatter() { }
if (false) {
    /** @type {?|undefined} */
    CustomDataFormatter.prototype.rowStyler;
    /** @type {?|undefined} */
    CustomDataFormatter.prototype.cellStyler;
    /* Skipping unhandled member: [fieldName: string]: any ;*/
}
