// @edp-bif/runtime-api/esm2015/lib/entity/form-rest/FormBillCreationProcessExecutorResult.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/form-rest/FormBillCreationProcessExecutorResult.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { JsonConvertableImpl, BpmnModelJsonHelper } from "@edp-aif/common-api";
import { ConditionResult } from '../ConditionResult';
import { RpcCreationRuleExecutorResult } from '@edp-aif/runtime-api';
export class FormBillCreationProcessExecutorResult extends JsonConvertableImpl {
    /**
     * @protected
     * @return {?}
     */
    ResetBeforeLoadFrom() {
        super.ResetBeforeLoadFrom();
        this.conditionResult = null;
        this.creationRuleExecutorResult = null;
    }
    /**
     * @return {?}
     */
    ConvertToJsonObject() {
        /** @type {?} */
        let obj = super.ConvertToJsonObject();
        BpmnModelJsonHelper.SetAttributeValue(obj, "conditionResult", this.conditionResult);
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleExecutorResult", this.creationRuleExecutorResult);
        return obj;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    LoadFromJsonObject(obj) {
        super.LoadFromJsonObject(obj);
        this.conditionResult = BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "conditionResult", ConditionResult);
        this.creationRuleExecutorResult = BpmnModelJsonHelper.GetAttributeValueAsJsonSerializable(obj, "creationRuleExecutorResult", RpcCreationRuleExecutorResult);
    }
}
if (false) {
    /** @type {?} */
    FormBillCreationProcessExecutorResult.prototype.conditionResult;
    /** @type {?} */
    FormBillCreationProcessExecutorResult.prototype.creationRuleExecutorResult;
}
