// @farris/ui-list-nav/esm5/lib/list-nav.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Directive, TemplateRef, ContentChild, Output, EventEmitter } from '@angular/core';
var listNavHeaderTemplateDirective = /** @class */ (function () {
    function listNavHeaderTemplateDirective(template) {
        this.template = template;
    }
    listNavHeaderTemplateDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[listNavHeader]'
                },] }
    ];
    /** @nocollapse */
    listNavHeaderTemplateDirective.ctorParameters = function () { return [
        { type: TemplateRef }
    ]; };
    return listNavHeaderTemplateDirective;
}());
export { listNavHeaderTemplateDirective };
if (false) {
    /** @type {?} */
    listNavHeaderTemplateDirective.prototype.template;
}
var listNavContentTemplateDirective = /** @class */ (function () {
    function listNavContentTemplateDirective(template) {
        this.template = template;
    }
    listNavContentTemplateDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[listNavContent]'
                },] }
    ];
    /** @nocollapse */
    listNavContentTemplateDirective.ctorParameters = function () { return [
        { type: TemplateRef }
    ]; };
    return listNavContentTemplateDirective;
}());
export { listNavContentTemplateDirective };
if (false) {
    /** @type {?} */
    listNavContentTemplateDirective.prototype.template;
}
var listNavFooterTemplateDirective = /** @class */ (function () {
    function listNavFooterTemplateDirective(template) {
        this.template = template;
    }
    listNavFooterTemplateDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[listNavFooter]'
                },] }
    ];
    /** @nocollapse */
    listNavFooterTemplateDirective.ctorParameters = function () { return [
        { type: TemplateRef }
    ]; };
    return listNavFooterTemplateDirective;
}());
export { listNavFooterTemplateDirective };
if (false) {
    /** @type {?} */
    listNavFooterTemplateDirective.prototype.template;
}
var ListNavComponent = /** @class */ (function () {
    /**是否展开navlist */
    // showList:boolean = true;
    function ListNavComponent() {
        /**
         * listnav宽度
         */
        this.listNavWidth = 218;
        // 显示入口
        this.showEntry = true;
        this.hover = false;
        this.disabled = false;
        this.position = 'left';
        this.hideNav = false;
        /**link发生变化 */
        // @Output('link-change') linkChange = new EventEmitter();
        /**
         * 收起折叠状态变化
         */
        this.navStateChange = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ListNavComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @param {?} event
     * @return {?}
     */
    ListNavComponent.prototype.toggle = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event.stopPropagation();
        if (this.disabled) {
            return;
        }
        this.hideNav = !this.hideNav;
        this.navStateChange.emit({ 'showNav': !this.hideNav });
    };
    /**
     * @param {?} event
     * @return {?}
     */
    ListNavComponent.prototype.hoverToggle = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.hideNav = !this.hideNav;
        this.navStateChange.emit({ 'showNav': !this.hideNav });
    };
    ListNavComponent.decorators = [
        { type: Component, args: [{
                    selector: 'farris-list-nav',
                    template: "<div class=\"f-list-nav\" [ngClass]=\"{'f-list-nav-left': position === 'left','f-list-nav-right': position === 'right'}\">\r\n    <div class=\"f-list-nav-in\" [ngStyle]=\"{'width': !hideNav? listNavWidth +'px' : '0px','transition':'width .5s'}\">\r\n        <div class=\"f-list-nav-main\" [hidden]=\"hideNav\">\r\n            <div class=\"f-list-nav-header\" *ngIf=\"HeaderTemplate\">\r\n                <ng-container [ngTemplateOutlet]=\"HeaderTemplate\"></ng-container>\r\n            </div>\r\n            <div class=\"f-list-nav-header\" *ngIf=\"!HeaderTemplate && title\">\r\n                <div class=\"f-list-nav-title\">\r\n                    {{title}}\r\n                </div>\r\n            </div>\r\n            <div class=\"f-list-nav-content\" *ngIf=\"NavContentTemplate\">\r\n                <ng-container [ngTemplateOutlet]=\"NavContentTemplate\"></ng-container>\r\n            </div>\r\n            <div class=\"f-list-nav-footer\" *ngIf=\"NavFooterTemplate\">\r\n                <ng-container [ngTemplateOutlet]=\"NavFooterTemplate\"></ng-container>\r\n            </div>\r\n        </div>\r\n        <div class=\"f-list-nav-toggle-sidebar\" *ngIf=\"showEntry\" [ngClass]=\"{'f-list-nav-toggle-disabled': disabled,'active': hideNav}\" (click)=\"toggle($event)\">\r\n            <span class=\"triangle\"></span>\r\n        </div>\r\n    </div>\r\n    \r\n</div>",
                    styles: [".f-list-nav{height:100%;min-height:200px}.f-list-nav .f-list-nav-in{position:relative;height:100%;background-color:#fff;box-shadow:1px 1px 8px 0 rgba(0,28,64,.08)}.f-list-nav .f-list-nav-main{display:-webkit-box;display:flex;-webkit-box-orient:vertical;-webkit-box-direction:normal;flex-direction:column;height:100%;overflow-x:hidden}.f-list-nav .f-list-nav-main .f-list-nav-footer,.f-list-nav .f-list-nav-main .f-list-nav-header{flex-shrink:0}.f-list-nav .f-list-nav-main .f-list-nav-content{-webkit-box-flex:1;flex-grow:1;flex-shrink:1;flex-basis:0;overflow-y:auto}.f-list-nav .f-list-nav-main .f-list-nav-header .f-list-nav-title{padding:1rem 3rem 1rem 2.5rem;line-height:1.375rem;font-size:1rem;color:#000;border-bottom:1px solid #e9e9e9;white-space:nowrap;overflow:hidden;text-overflow:ellipsis}.f-list-nav .f-list-nav-toggle-sidebar{position:absolute;right:-.375rem;top:50%;z-index:99;margin-top:-2.375rem;width:.375rem;height:4.75rem;line-height:4.75rem;text-align:center;border-radius:0 4px 4px 0;cursor:pointer}.f-list-nav .f-list-nav-toggle-sidebar .triangle{display:inline-block;width:0;height:0}.f-list-nav .f-list-nav-toggle-sidebar.f-list-nav-toggle-disabled{cursor:not-allowed}.f-list-nav .f-list-nav-toggle-sidebar.f-list-nav-toggle-disabled:hover{background:rgba(0,28,64,.1)}.f-list-nav .f-list-nav-toggle-sidebar.f-list-nav-toggle-disabled:hover .triangle{border-right:.25rem solid #818fab}.f-list-nav.f-list-nav-left{padding-right:.375rem}.f-list-nav.f-list-nav-left .f-list-nav-toggle-sidebar .triangle{border-top:.3875rem solid transparent;border-bottom:.3875rem solid transparent;border-right:.25rem solid #818fab}.f-list-nav.f-list-nav-left .f-list-nav-toggle-sidebar:hover .triangle{border-right:.25rem solid #fff}.f-list-nav.f-list-nav-left .f-list-nav-toggle-sidebar.active .triangle{border-right:none;border-left:.25rem solid #fff}.f-list-nav.f-list-nav-right{padding-left:.375rem}.f-list-nav.f-list-nav-right .f-list-nav-toggle-sidebar{left:-.375rem}.f-list-nav.f-list-nav-right .f-list-nav-toggle-sidebar .triangle{border-top:.3875rem solid transparent;border-bottom:.3875rem solid transparent;border-left:.25rem solid #818fab}.f-list-nav.f-list-nav-right .f-list-nav-toggle-sidebar:hover .triangle{border-left:.25rem solid #fff}.f-list-nav.f-list-nav-right .f-list-nav-toggle-sidebar.active .triangle{border-left:none;border-right:.25rem solid #fff}"]
                }] }
    ];
    /** @nocollapse */
    ListNavComponent.ctorParameters = function () { return []; };
    ListNavComponent.propDecorators = {
        HeaderTemplate: [{ type: ContentChild, args: [listNavHeaderTemplateDirective, { read: TemplateRef },] }],
        NavContentTemplate: [{ type: ContentChild, args: [listNavContentTemplateDirective, { read: TemplateRef },] }],
        NavFooterTemplate: [{ type: ContentChild, args: [listNavFooterTemplateDirective, { read: TemplateRef },] }],
        title: [{ type: Input }],
        listNavWidth: [{ type: Input }],
        showEntry: [{ type: Input }],
        hover: [{ type: Input }],
        disabled: [{ type: Input }],
        position: [{ type: Input }],
        hideNav: [{ type: Input }],
        navStateChange: [{ type: Output, args: ['navStateChange',] }]
    };
    return ListNavComponent;
}());
export { ListNavComponent };
if (false) {
    /** @type {?} */
    ListNavComponent.prototype.HeaderTemplate;
    /** @type {?} */
    ListNavComponent.prototype.NavContentTemplate;
    /** @type {?} */
    ListNavComponent.prototype.NavFooterTemplate;
    /**
     * listnav 名称
     * @type {?}
     */
    ListNavComponent.prototype.title;
    /**
     * listnav宽度
     * @type {?}
     */
    ListNavComponent.prototype.listNavWidth;
    /** @type {?} */
    ListNavComponent.prototype.showEntry;
    /** @type {?} */
    ListNavComponent.prototype.hover;
    /** @type {?} */
    ListNavComponent.prototype.disabled;
    /** @type {?} */
    ListNavComponent.prototype.position;
    /** @type {?} */
    ListNavComponent.prototype.hideNav;
    /**
     * 收起折叠状态变化
     * @type {?}
     */
    ListNavComponent.prototype.navStateChange;
}
