// @edp-bif/common-api/esm5/lib/entity/impl/ConditionEntity.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/ConditionEntity.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { JsonConvertableImpl, BpmnModelJsonHelper } from '../../ref';
import { ConditionType } from '../ConditionType';
/**
 * 条件实体类。
 */
var /**
 * 条件实体类。
 */
ConditionEntity = /** @class */ (function (_super) {
    tslib_1.__extends(ConditionEntity, _super);
    function ConditionEntity() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.isEnabled = false;
        _this.remarks = null;
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    ConditionEntity.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.id = null;
        this.isEnabled = false;
        this.name = null;
        this.type = null;
        this.content = null;
        this.remarks = null;
    };
    /**
     * @return {?}
     */
    ConditionEntity.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "id", this.id);
        BpmnModelJsonHelper.SetAttributeValue(obj, "isEnabled", this.isEnabled);
        BpmnModelJsonHelper.SetAttributeValue(obj, "name", this.name);
        BpmnModelJsonHelper.SetAttributeValue(obj, "type", this.type);
        BpmnModelJsonHelper.SetAttributeValue(obj, "content", this.content);
        BpmnModelJsonHelper.SetAttributeValue(obj, "remarks", this.remarks);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    ConditionEntity.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.id = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "id");
        this.isEnabled = BpmnModelJsonHelper.GetAttributeValueAsBoolean(obj, "isEnabled");
        this.name = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "name");
        this.type = ConditionType[BpmnModelJsonHelper.GetAttributeValueAsString(obj, "type")];
        this.content = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "content");
        this.remarks = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "remarks");
    };
    return ConditionEntity;
}(JsonConvertableImpl));
/**
 * 条件实体类。
 */
export { ConditionEntity };
if (false) {
    /** @type {?} */
    ConditionEntity.prototype.id;
    /** @type {?} */
    ConditionEntity.prototype.isEnabled;
    /** @type {?} */
    ConditionEntity.prototype.name;
    /** @type {?} */
    ConditionEntity.prototype.type;
    /** @type {?} */
    ConditionEntity.prototype.content;
    /** @type {?} */
    ConditionEntity.prototype.remarks;
}
