// @farris/ui-common/esm2015/lib/service/runtime-state.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * @Author: 疯狂秀才(Lucas Huang)
 * @Date: 2019-06-16 13:44:59
 * @LastEditors: 疯狂秀才(Lucas Huang)
 * @LastEditTime: 2019-08-15 09:47:34
 * @QQ: 1055818239
 * @Version: v0.0.1
 */
import { BehaviorSubject } from 'rxjs';
import { Injectable } from '@angular/core';
import { merge } from 'lodash-es';
import { map } from 'rxjs/operators';
import { CommonUtils } from './../common.utils';
import * as i0 from "@angular/core";
import * as i1 from "../common.utils";
/**
 * @record
 */
export function RuntimeState() { }
if (false) {
    /** @type {?|undefined} */
    RuntimeState.prototype.form;
    /** @type {?|undefined} */
    RuntimeState.prototype.model;
}
/** @type {?} */
const DefaultRuntimeState = {
    form: {},
    model: {}
};
export class RuntimeStateService {
    /**
     * @param {?} utils
     */
    constructor(utils) {
        this.utils = utils;
        this.stateSubject = new BehaviorSubject({});
        this.state$ = this.stateSubject.asObservable();
        this.form$ = this.state$.pipe(map((/**
         * @param {?} state
         * @return {?}
         */
        (state) => state.form)));
        this.lookupIsPending$ = this.form$.pipe(map((/**
         * @param {?} f
         * @return {?}
         */
        f => f.lookup.pending)));
        this._state = DefaultRuntimeState;
        this._formState = this._state.form;
        this._lookupState = this._formState.lookup;
    }
    /**
     * @private
     * @param {?} newVal
     * @return {?}
     */
    setValue(newVal) {
        if (newVal) {
            this._state = merge(this._state, newVal);
            this.stateSubject.next(this._state);
        }
    }
    /**
     * @private
     * @param {?} keyPath
     * @return {?}
     */
    getValue(keyPath) {
        return this.utils.getValue(keyPath, this._state);
    }
    /**
     * @param {?} el
     * @return {?}
     */
    setLookupInstance(el) {
        this.setValue({ form: { lookup: { instance: el } } });
    }
    /**
     * @param {?} newVal
     * @return {?}
     */
    updateFormState(newVal) {
        this.setValue({ form: newVal });
    }
    /**
     * @param {?} keyPath
     * @return {?}
     */
    getFormState(keyPath) {
        return this.getValue('form.' + keyPath);
    }
    /**
     * @param {?} evt
     * @return {?}
     */
    eventPath(evt) {
        /** @type {?} */
        const path = (evt.composedPath && evt.composedPath()) || evt.path;
        /** @type {?} */
        const target = evt.target;
        if (path != null) {
            return (path.indexOf(window) < 0) ? path.concat(window) : path;
        }
        if (target === window) {
            return [window];
        }
        /** @type {?} */
        const getParents = (/**
         * @param {?} node
         * @param {?=} memo
         * @return {?}
         */
        (node, memo = undefined) => {
            memo = memo || [];
            /** @type {?} */
            const parentNode = node.parentNode;
            if (!parentNode) {
                return memo;
            }
            else {
                return getParents(parentNode, memo.concat(parentNode));
            }
        });
        return [target].concat(getParents(target), window);
    }
}
RuntimeStateService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
RuntimeStateService.ctorParameters = () => [
    { type: CommonUtils }
];
/** @nocollapse */ RuntimeStateService.ngInjectableDef = i0.defineInjectable({ factory: function RuntimeStateService_Factory() { return new RuntimeStateService(i0.inject(i1.CommonUtils)); }, token: RuntimeStateService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    RuntimeStateService.prototype._state;
    /**
     * @type {?}
     * @private
     */
    RuntimeStateService.prototype.stateSubject;
    /** @type {?} */
    RuntimeStateService.prototype.state$;
    /** @type {?} */
    RuntimeStateService.prototype.form$;
    /** @type {?} */
    RuntimeStateService.prototype.lookupIsPending$;
    /**
     * @type {?}
     * @private
     */
    RuntimeStateService.prototype._formState;
    /**
     * @type {?}
     * @private
     */
    RuntimeStateService.prototype._lookupState;
    /**
     * @type {?}
     * @private
     */
    RuntimeStateService.prototype.utils;
}
