// @farris/ui-organization-selector/esm5/lib/organization-selector.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/organization-selector.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { InputGroupModule } from '@farris/ui-input-group';
import { TreeTableModule } from '@farris/ui-treetable';
import { FarrisDialogModule } from '@farris/ui-dialog';
import { FarrisTabsModule } from '@farris/ui-tabs';
import { PerfectScrollbarModule } from '@farris/ui-perfect-scrollbar';
import { MessagerModule } from "@farris/ui-messager";
import { AvatarModule } from '@farris/ui-avatar';
import { OrganizationSelectorComponent } from './organization-selector.component';
import { OrganizationComponent } from '../lib/tab-organization/to.component';
import { SelectionComponent } from '../lib/selections/sl.component';
import { SearchComponent } from './search/srp.component';
import { FavoritesComponent } from './tab-favorites/favorites.component';
import { GroupComponent } from './tab-group/group.component';
import { AppendGroupOrgComponent } from './append-group-org/appendorg.component';
import { CommandServicesModule } from '@farris/command-services';
var OrganizationSelectorModule = /** @class */ (function () {
    function OrganizationSelectorModule() {
    }
    OrganizationSelectorModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [OrganizationSelectorComponent, OrganizationComponent, SelectionComponent, SearchComponent, FavoritesComponent, GroupComponent, AppendGroupOrgComponent],
                    imports: [
                        CommonModule,
                        InputGroupModule,
                        TreeTableModule,
                        FarrisDialogModule,
                        FarrisTabsModule,
                        PerfectScrollbarModule,
                        FormsModule,
                        ReactiveFormsModule,
                        MessagerModule.forRoot(),
                        AvatarModule,
                        CommandServicesModule
                    ],
                    exports: [OrganizationSelectorComponent],
                    entryComponents: [OrganizationSelectorComponent],
                    providers: []
                },] }
    ];
    return OrganizationSelectorModule;
}());
export { OrganizationSelectorModule };
