// @farris/devkit/esm2015/lib/message/messagemanager.js
export class MessageManager {
    // private parent:Window;
    // private children = new Map<string,Window>();
    constructor() {
    }
    /**
     * 获取当前window上的MessageManager对象
     */
    static getMessager() {
        if (!window['igix_messagemanager']) {
            window['igix_messagemanager'] = new MessageManager();
        }
        return window['igix_messagemanager'];
    }
    // addParentTarget(target: Window){
    //     this.parent = target;
    // }
    // addChildTarget(key: string,target: Window){
    //     this.children.set(key,target);
    // }
    listner(callbackfn) {
        window.addEventListener('message', function (e) {
            const data = e.data;
            data.sender = e.source;
            callbackfn(data);
        });
    }
    // listnerMessage(targetId: string,callbackfn: (e: any) => any){
    //     const win = this.children.get(targetId);
    //     if(!win){
    //         throw Error('This TargetId Can Not Find Window Object');
    //     }
    //     win.addEventListener('message',async function(e){
    //         await callbackfn(e);
    //     });
    // }
    send(msg) {
        // const win = this.children.get(targetId);
        let win;
        if (msg.target) {
            win = msg.target;
            delete msg.target;
        }
        else {
            win = window;
        }
        win.postMessage(msg, '*');
    }
    setAppContext(appContext) {
        this.curAppContext = appContext;
    }
    getAppContext() {
        return this.curAppContext;
    }
}
